{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "source": [
        "pip install biopython"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TFC-9Ku-yHK8",
        "outputId": "620f68a4-ce3e-4a8c-c4a9-1edabd5ae8a0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.83-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m12.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.25.2)\n",
            "Installing collected packages: biopython\n",
            "Successfully installed biopython-1.83\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TxXNrjqByIo1",
        "outputId": "638f8041-4a9a-4e24-e7a5-bb2767c24dee"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Reading a DNA Sequence from a FASTA File using Biopython\n",
        "\n",
        "In this block of code, we are using the `SeqIO` module from Biopython to parse a DNA sequence from a FASTA file (`hxk.fasta`). The FASTA file is assumed to be located in Google Drive. We loop through the parsed file, and assign the first sequence record to the variable `hxk`, which stores it as a `SeqRecord` object. This object contains the sequence (`hxk.seq`), as well as metadata such as the sequence ID and description.\n",
        "\n",
        "hxk-HEXOKINASE\n"
      ],
      "metadata": {
        "id": "kfdhtNvI05D8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "for sequence in SeqIO.parse(\"/content/drive/MyDrive/EXAMPLE/hxk.fasta\",'fasta'):\n",
        "  hxk=sequence"
      ],
      "metadata": {
        "id": "jehwF6GISYRf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "print(hxk.seq)  # Print the DNA sequence\n",
        "print(hxk.seq.transcribe())  # Print the RNA transcript of the DNA sequence\n",
        "print(hxk.seq.reverse_complement())  # Print the reverse complement of the DNA sequence"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HsArJZgPS3mM",
        "outputId": "8ebffbc2-5c88-4a11-b368-18f360011c66"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ATCTATCTTGCTGTGTTTGGACAGGCCAGCCCCTGAAACATCTTGGGCAATGGAGGGTTAACTTCTCAAAGTTTAATAGGCAAGACCAGCAACCATGCAACAAGGACTTCAACTAACCAACTAAAGAACTGTTCCCCAGAGCATTGTTCCTGAGAAGGAAAAGAGTCCAAACACCTACCCACACCTGCTTTGTGCCAAGAATCCACAGTTGGATTGCAAGGACAGTGCGTTCAAGACCCAGCTGTTGAGAGTAGAAAAGCAGAAGAAAGGACCCGAGGTCAGCAAGTGCCCTCCCCACAATGGGGCAGATCTGCCAGCGAGAATCGGCTACAGCAGCTGAAAAACCAAAACTTCATCTACTTGCTGAAAGTGAGATTGACAAGTATCTCTATGCCATGCGGCTCTCCGATGAAACTCTCATAGATATCATGACTCGCTTCAGGAAGGAGATGAAGAATGGCCTCTCCCGGGATTTTAATCCAACAGCCACAGTCAAGATGTTGCCAACATTCGTAAGGTCCATTCCTGATGGCTCTGAAAAGGGAGATTTCATTGCCCTGGATCTTGGTGGGTCTTCCTTTCGAATTCTGCGGGTGCAAGTGAATCATGAGAAAAACCAGAATGTTCACATGGAGTCCGAGGTTTATGACACCCCAGAGAACATCGTGCACGGCAGTGGAAGCCAGCTTTTTGATCATGTTGCTGAGTGCCTGGGAGATTTCATGGAGAAAAGGAAGATCAAGGACAAGAAGTTACCTGTGGGATTCACGTTTTCTTTTCCTTGCCAACAATCCAAAATAGATGAGGCCATCCTGATCACCTGGACAAAGCGATTTAAAGCGAGCGGAGTGGAAGGAGCAGATGTGGTCAAACTGCTTAACAAAGCCATCAAAAAGCGAGGGGACTATGATGCCAACATCGTAGCTGTGGTGAATGACACAGTGGGCACCATGATGACCTGTGGCTATGACGACCAGCACTGTGAAGTCGGCCTGATCATCGGCACTGGCACCAATGCTTGCTACATGGAGGAACTGAGGCACATTGATCTGGTGGAAGGAGACGAGGGGAGGATGTGTATCAATACAGAATGGGGAGCCTTTGGAGACGATGGATCATTAGAAGACATCCGGACAGAGTTTGACAGGGAGATAGACCGGGGATCCCTCAACCCTGGAAAACAGCTGTTTGAGAAGATGGTCAGTGGCATGTACTTGGGAGAGCTGGTTCGACTGATCCTAGTCAAGATGGCCAAGGAGGGCCTCTTATTTGAAGGGCGGATCACCCCGGAGCTGCTCACCCGAGGGAAGTTTAACACCAGTGATGTGTCAGCCATCGAAAAGAATAAGGAAGGCCTCCACAATGCCAAAGAAATCCTGACCCGCCTGGGAGTGGAGCCGTCCGATGATGACTGTGTCTCAGTCCAGCACGTTTGCACCATTGTCTCATTTCGCTCAGCCAACTTGGTGGCTGCCACACTGGGCGCCATCTTGAACCGCCTGCGTGATAACAAGGGCACACCCAGGCTGCGGACCACGGTTGGTGTCGACGGATCTCTTTACAAGACGCACCCACAGTATTCCCGGCGTTTCCACAAGACTCTAAGGCGCTTGGTGCCAGACTCCGATGTGCGCTTCCTCCTCTCGGAGAGTGGCAGCGGCAAGGGGGCTGCCATGGTGACGGCGGTGGCCTACCGCTTGGCCGAGCAGCACCGGCAGATAGAGGAGACCCTGGCTCATTTCCACCTCACCAAGGACATGCTGCTGGAGGTGAAGAAGAGGATGCGGGCCGAGATGGAGCTGGGGCTGAGGAAGCAGACGCACAACAATGCCGTGGTTAAGATGCTGCCCTCCTTCGTCCGGAGAACTCCCGACGGGACCGAGAATGGTGACTTCTTGGCCCTGGATCTTGGAGGAACCAATTTCCGTGTGCTGCTGGTGAAAATCCGTAGTGGGAAAAAGAGAACGGTGGAAATGCACAACAAGATCTACGCCATTCCTATTGAAATCATGCAGGGCACTGGGGAAGAGCTGTTTGATCACATTGTCTCCTGCATCTCTGACTTCTTGGACTACATGGGGATCAAAGGCCCCAGGATGCCTCTGGGCTTCACGTTCTCATTTCCCTGCCAGCAGACGAGTCTGGACGCGGGAATCTTGATCACGTGGACAAAGGGTTTTAAGGCAACAGACTGCGTGGGCCACGATGTAGTCACCTTACTAAGGGATGCGATAAAAAGGAGAGAGGAATTTGACCTGGACGTGGTGGCTGTGGTCAACGACACAGTGGGCACCATGATGACCTGTGCTTATGAGGAGCCCACCTGTGAGGTTGGACTCATTGTTGGGACCGGCAGCAATGCCTGCTACATGGAGGAGATGAAGAACGTGGAGATGGTGGAGGGGGACCAGGGGCAGATGTGCATCAACATGGAGTGGGGGGCCTTTGGGGACAACGGGTGTCTGGATGATATCAGGACACACTACGACAGACTGGTGGACGAATATTCCCTAAATGCTGGGAAACAAAGGTATGAGAAGATGATCAGTGGTATGTACCTGGGTGAAATCGTCCGCAACATCTTAATCGACTTCACCAAGAAGGGATTCCTCTTCCGAGGGCAGATCTCTGAGACGCTGAAGACCCGGGGCATCTTTGAGACCAAGTTTCTCTCTCAGATCGAGAGTGACCGATTAGCACTGCTCCAGGTCCGGGCTATCCTCCAGCAGCTAGGTCTGAATAGCACCTGCGATGACAGTATCCTCGTCAAGACAGTGTGCGGGGTGGTGTCCAGGAGGGCCGCACAGCTGTGTGGCGCAGGCATGGCTGCGGTTGTGGATAAGATCCGCGAGAACAGAGGACTGGACCGTCTGAATGTGACTGTGGGAGTGGACGGGACACTCTACAAGCTTCATCCACACTTCTCCAGAATCATGCACCAGACGGTGAAGGAACTGTCACCAAAATGTAACGTGTCCTTCCTCCTGTCTGAGGATGGCAGCGGCAAGGGGGCCGCCCTCATCACGGCCGTGGGCGTGCGGTTACGCACAGAGGCAAGCAGCTAAGAGTCCGGGATCCCCAGCCTACTGCCTCTCCAGCACTTCTCTCTTCAAGCGGCGACCCCCTACCCTCCCAGCGAGTTGCGCTGGGAGACGCTGGCGCCAGGGCCTGCCGGCGCGGGGAGGAAAGCAAAATCCAACTAATGGTATATATTGTAGGGTACAGAATAGAGCGTGTGCTGTTGATAATATCTCTCACCCGGATCCCTCCTCACTTGCCCTGCCACTTTGCATGGTTTGATTTTGACCTGGTCCCCCACGTGTGAAGTGTAGTGGCATCCATTTCTAATGTATGCATTCATCCAACAGAGTTATTTATTGGCTGGAGATGGAAAATCACACCACCTGACAGGCCTTCTGGGCCTCCAAAGCCCATCCTTGGGGTTCCCCCTCCCTGTGTGAAATGTATTATCACCAGCAGACACTGCCGGGCCTCCCTCCCGGGGGCACTGCCTGAAGGCGAGTGTGGGCATAGCATTAGCTGCTTCCTCCCCTCCTGGCACCCACTGTGGCCTGGCATCGCATCGTGGTGTGTCAATGCCACAAAATCGTGTGTCCGTGGAACCAGTCCTAGCCGCGTGTGACAGTCTTGCATTCTGTTTGTCTCGTGGGGGGAGGTGGACAGTCCTGCGGAAATGTGTCTTGTCTCCATTTGGATAAAAGGAACCAACCAACAAACAATGCCATCACTGGAATTTCCCACCGCTTTGTGAGCCGTGTCGTATGACCTAGTAAACTTTGTACCAATTCAAA\n",
            "AUCUAUCUUGCUGUGUUUGGACAGGCCAGCCCCUGAAACAUCUUGGGCAAUGGAGGGUUAACUUCUCAAAGUUUAAUAGGCAAGACCAGCAACCAUGCAACAAGGACUUCAACUAACCAACUAAAGAACUGUUCCCCAGAGCAUUGUUCCUGAGAAGGAAAAGAGUCCAAACACCUACCCACACCUGCUUUGUGCCAAGAAUCCACAGUUGGAUUGCAAGGACAGUGCGUUCAAGACCCAGCUGUUGAGAGUAGAAAAGCAGAAGAAAGGACCCGAGGUCAGCAAGUGCCCUCCCCACAAUGGGGCAGAUCUGCCAGCGAGAAUCGGCUACAGCAGCUGAAAAACCAAAACUUCAUCUACUUGCUGAAAGUGAGAUUGACAAGUAUCUCUAUGCCAUGCGGCUCUCCGAUGAAACUCUCAUAGAUAUCAUGACUCGCUUCAGGAAGGAGAUGAAGAAUGGCCUCUCCCGGGAUUUUAAUCCAACAGCCACAGUCAAGAUGUUGCCAACAUUCGUAAGGUCCAUUCCUGAUGGCUCUGAAAAGGGAGAUUUCAUUGCCCUGGAUCUUGGUGGGUCUUCCUUUCGAAUUCUGCGGGUGCAAGUGAAUCAUGAGAAAAACCAGAAUGUUCACAUGGAGUCCGAGGUUUAUGACACCCCAGAGAACAUCGUGCACGGCAGUGGAAGCCAGCUUUUUGAUCAUGUUGCUGAGUGCCUGGGAGAUUUCAUGGAGAAAAGGAAGAUCAAGGACAAGAAGUUACCUGUGGGAUUCACGUUUUCUUUUCCUUGCCAACAAUCCAAAAUAGAUGAGGCCAUCCUGAUCACCUGGACAAAGCGAUUUAAAGCGAGCGGAGUGGAAGGAGCAGAUGUGGUCAAACUGCUUAACAAAGCCAUCAAAAAGCGAGGGGACUAUGAUGCCAACAUCGUAGCUGUGGUGAAUGACACAGUGGGCACCAUGAUGACCUGUGGCUAUGACGACCAGCACUGUGAAGUCGGCCUGAUCAUCGGCACUGGCACCAAUGCUUGCUACAUGGAGGAACUGAGGCACAUUGAUCUGGUGGAAGGAGACGAGGGGAGGAUGUGUAUCAAUACAGAAUGGGGAGCCUUUGGAGACGAUGGAUCAUUAGAAGACAUCCGGACAGAGUUUGACAGGGAGAUAGACCGGGGAUCCCUCAACCCUGGAAAACAGCUGUUUGAGAAGAUGGUCAGUGGCAUGUACUUGGGAGAGCUGGUUCGACUGAUCCUAGUCAAGAUGGCCAAGGAGGGCCUCUUAUUUGAAGGGCGGAUCACCCCGGAGCUGCUCACCCGAGGGAAGUUUAACACCAGUGAUGUGUCAGCCAUCGAAAAGAAUAAGGAAGGCCUCCACAAUGCCAAAGAAAUCCUGACCCGCCUGGGAGUGGAGCCGUCCGAUGAUGACUGUGUCUCAGUCCAGCACGUUUGCACCAUUGUCUCAUUUCGCUCAGCCAACUUGGUGGCUGCCACACUGGGCGCCAUCUUGAACCGCCUGCGUGAUAACAAGGGCACACCCAGGCUGCGGACCACGGUUGGUGUCGACGGAUCUCUUUACAAGACGCACCCACAGUAUUCCCGGCGUUUCCACAAGACUCUAAGGCGCUUGGUGCCAGACUCCGAUGUGCGCUUCCUCCUCUCGGAGAGUGGCAGCGGCAAGGGGGCUGCCAUGGUGACGGCGGUGGCCUACCGCUUGGCCGAGCAGCACCGGCAGAUAGAGGAGACCCUGGCUCAUUUCCACCUCACCAAGGACAUGCUGCUGGAGGUGAAGAAGAGGAUGCGGGCCGAGAUGGAGCUGGGGCUGAGGAAGCAGACGCACAACAAUGCCGUGGUUAAGAUGCUGCCCUCCUUCGUCCGGAGAACUCCCGACGGGACCGAGAAUGGUGACUUCUUGGCCCUGGAUCUUGGAGGAACCAAUUUCCGUGUGCUGCUGGUGAAAAUCCGUAGUGGGAAAAAGAGAACGGUGGAAAUGCACAACAAGAUCUACGCCAUUCCUAUUGAAAUCAUGCAGGGCACUGGGGAAGAGCUGUUUGAUCACAUUGUCUCCUGCAUCUCUGACUUCUUGGACUACAUGGGGAUCAAAGGCCCCAGGAUGCCUCUGGGCUUCACGUUCUCAUUUCCCUGCCAGCAGACGAGUCUGGACGCGGGAAUCUUGAUCACGUGGACAAAGGGUUUUAAGGCAACAGACUGCGUGGGCCACGAUGUAGUCACCUUACUAAGGGAUGCGAUAAAAAGGAGAGAGGAAUUUGACCUGGACGUGGUGGCUGUGGUCAACGACACAGUGGGCACCAUGAUGACCUGUGCUUAUGAGGAGCCCACCUGUGAGGUUGGACUCAUUGUUGGGACCGGCAGCAAUGCCUGCUACAUGGAGGAGAUGAAGAACGUGGAGAUGGUGGAGGGGGACCAGGGGCAGAUGUGCAUCAACAUGGAGUGGGGGGCCUUUGGGGACAACGGGUGUCUGGAUGAUAUCAGGACACACUACGACAGACUGGUGGACGAAUAUUCCCUAAAUGCUGGGAAACAAAGGUAUGAGAAGAUGAUCAGUGGUAUGUACCUGGGUGAAAUCGUCCGCAACAUCUUAAUCGACUUCACCAAGAAGGGAUUCCUCUUCCGAGGGCAGAUCUCUGAGACGCUGAAGACCCGGGGCAUCUUUGAGACCAAGUUUCUCUCUCAGAUCGAGAGUGACCGAUUAGCACUGCUCCAGGUCCGGGCUAUCCUCCAGCAGCUAGGUCUGAAUAGCACCUGCGAUGACAGUAUCCUCGUCAAGACAGUGUGCGGGGUGGUGUCCAGGAGGGCCGCACAGCUGUGUGGCGCAGGCAUGGCUGCGGUUGUGGAUAAGAUCCGCGAGAACAGAGGACUGGACCGUCUGAAUGUGACUGUGGGAGUGGACGGGACACUCUACAAGCUUCAUCCACACUUCUCCAGAAUCAUGCACCAGACGGUGAAGGAACUGUCACCAAAAUGUAACGUGUCCUUCCUCCUGUCUGAGGAUGGCAGCGGCAAGGGGGCCGCCCUCAUCACGGCCGUGGGCGUGCGGUUACGCACAGAGGCAAGCAGCUAAGAGUCCGGGAUCCCCAGCCUACUGCCUCUCCAGCACUUCUCUCUUCAAGCGGCGACCCCCUACCCUCCCAGCGAGUUGCGCUGGGAGACGCUGGCGCCAGGGCCUGCCGGCGCGGGGAGGAAAGCAAAAUCCAACUAAUGGUAUAUAUUGUAGGGUACAGAAUAGAGCGUGUGCUGUUGAUAAUAUCUCUCACCCGGAUCCCUCCUCACUUGCCCUGCCACUUUGCAUGGUUUGAUUUUGACCUGGUCCCCCACGUGUGAAGUGUAGUGGCAUCCAUUUCUAAUGUAUGCAUUCAUCCAACAGAGUUAUUUAUUGGCUGGAGAUGGAAAAUCACACCACCUGACAGGCCUUCUGGGCCUCCAAAGCCCAUCCUUGGGGUUCCCCCUCCCUGUGUGAAAUGUAUUAUCACCAGCAGACACUGCCGGGCCUCCCUCCCGGGGGCACUGCCUGAAGGCGAGUGUGGGCAUAGCAUUAGCUGCUUCCUCCCCUCCUGGCACCCACUGUGGCCUGGCAUCGCAUCGUGGUGUGUCAAUGCCACAAAAUCGUGUGUCCGUGGAACCAGUCCUAGCCGCGUGUGACAGUCUUGCAUUCUGUUUGUCUCGUGGGGGGAGGUGGACAGUCCUGCGGAAAUGUGUCUUGUCUCCAUUUGGAUAAAAGGAACCAACCAACAAACAAUGCCAUCACUGGAAUUUCCCACCGCUUUGUGAGCCGUGUCGUAUGACCUAGUAAACUUUGUACCAAUUCAAA\n",
            "TTTGAATTGGTACAAAGTTTACTAGGTCATACGACACGGCTCACAAAGCGGTGGGAAATTCCAGTGATGGCATTGTTTGTTGGTTGGTTCCTTTTATCCAAATGGAGACAAGACACATTTCCGCAGGACTGTCCACCTCCCCCCACGAGACAAACAGAATGCAAGACTGTCACACGCGGCTAGGACTGGTTCCACGGACACACGATTTTGTGGCATTGACACACCACGATGCGATGCCAGGCCACAGTGGGTGCCAGGAGGGGAGGAAGCAGCTAATGCTATGCCCACACTCGCCTTCAGGCAGTGCCCCCGGGAGGGAGGCCCGGCAGTGTCTGCTGGTGATAATACATTTCACACAGGGAGGGGGAACCCCAAGGATGGGCTTTGGAGGCCCAGAAGGCCTGTCAGGTGGTGTGATTTTCCATCTCCAGCCAATAAATAACTCTGTTGGATGAATGCATACATTAGAAATGGATGCCACTACACTTCACACGTGGGGGACCAGGTCAAAATCAAACCATGCAAAGTGGCAGGGCAAGTGAGGAGGGATCCGGGTGAGAGATATTATCAACAGCACACGCTCTATTCTGTACCCTACAATATATACCATTAGTTGGATTTTGCTTTCCTCCCCGCGCCGGCAGGCCCTGGCGCCAGCGTCTCCCAGCGCAACTCGCTGGGAGGGTAGGGGGTCGCCGCTTGAAGAGAGAAGTGCTGGAGAGGCAGTAGGCTGGGGATCCCGGACTCTTAGCTGCTTGCCTCTGTGCGTAACCGCACGCCCACGGCCGTGATGAGGGCGGCCCCCTTGCCGCTGCCATCCTCAGACAGGAGGAAGGACACGTTACATTTTGGTGACAGTTCCTTCACCGTCTGGTGCATGATTCTGGAGAAGTGTGGATGAAGCTTGTAGAGTGTCCCGTCCACTCCCACAGTCACATTCAGACGGTCCAGTCCTCTGTTCTCGCGGATCTTATCCACAACCGCAGCCATGCCTGCGCCACACAGCTGTGCGGCCCTCCTGGACACCACCCCGCACACTGTCTTGACGAGGATACTGTCATCGCAGGTGCTATTCAGACCTAGCTGCTGGAGGATAGCCCGGACCTGGAGCAGTGCTAATCGGTCACTCTCGATCTGAGAGAGAAACTTGGTCTCAAAGATGCCCCGGGTCTTCAGCGTCTCAGAGATCTGCCCTCGGAAGAGGAATCCCTTCTTGGTGAAGTCGATTAAGATGTTGCGGACGATTTCACCCAGGTACATACCACTGATCATCTTCTCATACCTTTGTTTCCCAGCATTTAGGGAATATTCGTCCACCAGTCTGTCGTAGTGTGTCCTGATATCATCCAGACACCCGTTGTCCCCAAAGGCCCCCCACTCCATGTTGATGCACATCTGCCCCTGGTCCCCCTCCACCATCTCCACGTTCTTCATCTCCTCCATGTAGCAGGCATTGCTGCCGGTCCCAACAATGAGTCCAACCTCACAGGTGGGCTCCTCATAAGCACAGGTCATCATGGTGCCCACTGTGTCGTTGACCACAGCCACCACGTCCAGGTCAAATTCCTCTCTCCTTTTTATCGCATCCCTTAGTAAGGTGACTACATCGTGGCCCACGCAGTCTGTTGCCTTAAAACCCTTTGTCCACGTGATCAAGATTCCCGCGTCCAGACTCGTCTGCTGGCAGGGAAATGAGAACGTGAAGCCCAGAGGCATCCTGGGGCCTTTGATCCCCATGTAGTCCAAGAAGTCAGAGATGCAGGAGACAATGTGATCAAACAGCTCTTCCCCAGTGCCCTGCATGATTTCAATAGGAATGGCGTAGATCTTGTTGTGCATTTCCACCGTTCTCTTTTTCCCACTACGGATTTTCACCAGCAGCACACGGAAATTGGTTCCTCCAAGATCCAGGGCCAAGAAGTCACCATTCTCGGTCCCGTCGGGAGTTCTCCGGACGAAGGAGGGCAGCATCTTAACCACGGCATTGTTGTGCGTCTGCTTCCTCAGCCCCAGCTCCATCTCGGCCCGCATCCTCTTCTTCACCTCCAGCAGCATGTCCTTGGTGAGGTGGAAATGAGCCAGGGTCTCCTCTATCTGCCGGTGCTGCTCGGCCAAGCGGTAGGCCACCGCCGTCACCATGGCAGCCCCCTTGCCGCTGCCACTCTCCGAGAGGAGGAAGCGCACATCGGAGTCTGGCACCAAGCGCCTTAGAGTCTTGTGGAAACGCCGGGAATACTGTGGGTGCGTCTTGTAAAGAGATCCGTCGACACCAACCGTGGTCCGCAGCCTGGGTGTGCCCTTGTTATCACGCAGGCGGTTCAAGATGGCGCCCAGTGTGGCAGCCACCAAGTTGGCTGAGCGAAATGAGACAATGGTGCAAACGTGCTGGACTGAGACACAGTCATCATCGGACGGCTCCACTCCCAGGCGGGTCAGGATTTCTTTGGCATTGTGGAGGCCTTCCTTATTCTTTTCGATGGCTGACACATCACTGGTGTTAAACTTCCCTCGGGTGAGCAGCTCCGGGGTGATCCGCCCTTCAAATAAGAGGCCCTCCTTGGCCATCTTGACTAGGATCAGTCGAACCAGCTCTCCCAAGTACATGCCACTGACCATCTTCTCAAACAGCTGTTTTCCAGGGTTGAGGGATCCCCGGTCTATCTCCCTGTCAAACTCTGTCCGGATGTCTTCTAATGATCCATCGTCTCCAAAGGCTCCCCATTCTGTATTGATACACATCCTCCCCTCGTCTCCTTCCACCAGATCAATGTGCCTCAGTTCCTCCATGTAGCAAGCATTGGTGCCAGTGCCGATGATCAGGCCGACTTCACAGTGCTGGTCGTCATAGCCACAGGTCATCATGGTGCCCACTGTGTCATTCACCACAGCTACGATGTTGGCATCATAGTCCCCTCGCTTTTTGATGGCTTTGTTAAGCAGTTTGACCACATCTGCTCCTTCCACTCCGCTCGCTTTAAATCGCTTTGTCCAGGTGATCAGGATGGCCTCATCTATTTTGGATTGTTGGCAAGGAAAAGAAAACGTGAATCCCACAGGTAACTTCTTGTCCTTGATCTTCCTTTTCTCCATGAAATCTCCCAGGCACTCAGCAACATGATCAAAAAGCTGGCTTCCACTGCCGTGCACGATGTTCTCTGGGGTGTCATAAACCTCGGACTCCATGTGAACATTCTGGTTTTTCTCATGATTCACTTGCACCCGCAGAATTCGAAAGGAAGACCCACCAAGATCCAGGGCAATGAAATCTCCCTTTTCAGAGCCATCAGGAATGGACCTTACGAATGTTGGCAACATCTTGACTGTGGCTGTTGGATTAAAATCCCGGGAGAGGCCATTCTTCATCTCCTTCCTGAAGCGAGTCATGATATCTATGAGAGTTTCATCGGAGAGCCGCATGGCATAGAGATACTTGTCAATCTCACTTTCAGCAAGTAGATGAAGTTTTGGTTTTTCAGCTGCTGTAGCCGATTCTCGCTGGCAGATCTGCCCCATTGTGGGGAGGGCACTTGCTGACCTCGGGTCCTTTCTTCTGCTTTTCTACTCTCAACAGCTGGGTCTTGAACGCACTGTCCTTGCAATCCAACTGTGGATTCTTGGCACAAAGCAGGTGTGGGTAGGTGTTTGGACTCTTTTCCTTCTCAGGAACAATGCTCTGGGGAACAGTTCTTTAGTTGGTTAGTTGAAGTCCTTGTTGCATGGTTGCTGGTCTTGCCTATTAAACTTTGAGAAGTTAACCCTCCATTGCCCAAGATGTTTCAGGGGCTGGCCTGTCCAAACACAGCAAGATAGAT\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "dummy=hxk.seq[:10] # Taking the first 10 nucleotides from the DNA sequence\n",
        "print(\"Coding strand:\",dummy) # Print the coding strand (the selected segment)\n",
        "print(\"Template strand:\",dummy.complement()) # Print the complementary template strand\n",
        "print(\"mRNA:\",dummy.transcribe()) # Print the mRNA transcript from the coding strand\n",
        "print(\"Reverse complement:\",dummy.reverse_complement()) # Print the reverse complement of the coding strand\n",
        "print(\"Reverse complement RNA:\",dummy.reverse_complement_rna()) # Print the reverse complement of the mRNA transcript"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "eydMODEzTraH",
        "outputId": "ad8264f7-a9dc-4523-c520-6cf201dec3ba"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Coding strand: ATCTATCTTG\n",
            "Template strand: TAGATAGAAC\n",
            "mRNA: AUCUAUCUUG\n",
            "Reverse complement: CAAGATAGAT\n",
            "Reverse complement RNA: CAAGAUAGAU\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Performing operations on the entire DNA sequence in hxk"
      ],
      "metadata": {
        "id": "fUvYYV4kJhsX"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dummy=hxk.seq\n",
        "print(\"Coding strand:\",dummy)\n",
        "print(\"Template strand:\",dummy.complement())\n",
        "print(\"mRNA:\",dummy.transcribe())\n",
        "print(\"Reverse complement:\",dummy.reverse_complement())\n",
        "print(\"Reverse complement RNA:\",dummy.reverse_complement_rna())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "J_NP8SZ3_MGJ",
        "outputId": "b6f8f592-f7f1-4f57-add9-8c7681835203"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Coding strand: ATCTATCTTGCTGTGTTTGGACAGGCCAGCCCCTGAAACATCTTGGGCAATGGAGGGTTAACTTCTCAAAGTTTAATAGGCAAGACCAGCAACCATGCAACAAGGACTTCAACTAACCAACTAAAGAACTGTTCCCCAGAGCATTGTTCCTGAGAAGGAAAAGAGTCCAAACACCTACCCACACCTGCTTTGTGCCAAGAATCCACAGTTGGATTGCAAGGACAGTGCGTTCAAGACCCAGCTGTTGAGAGTAGAAAAGCAGAAGAAAGGACCCGAGGTCAGCAAGTGCCCTCCCCACAATGGGGCAGATCTGCCAGCGAGAATCGGCTACAGCAGCTGAAAAACCAAAACTTCATCTACTTGCTGAAAGTGAGATTGACAAGTATCTCTATGCCATGCGGCTCTCCGATGAAACTCTCATAGATATCATGACTCGCTTCAGGAAGGAGATGAAGAATGGCCTCTCCCGGGATTTTAATCCAACAGCCACAGTCAAGATGTTGCCAACATTCGTAAGGTCCATTCCTGATGGCTCTGAAAAGGGAGATTTCATTGCCCTGGATCTTGGTGGGTCTTCCTTTCGAATTCTGCGGGTGCAAGTGAATCATGAGAAAAACCAGAATGTTCACATGGAGTCCGAGGTTTATGACACCCCAGAGAACATCGTGCACGGCAGTGGAAGCCAGCTTTTTGATCATGTTGCTGAGTGCCTGGGAGATTTCATGGAGAAAAGGAAGATCAAGGACAAGAAGTTACCTGTGGGATTCACGTTTTCTTTTCCTTGCCAACAATCCAAAATAGATGAGGCCATCCTGATCACCTGGACAAAGCGATTTAAAGCGAGCGGAGTGGAAGGAGCAGATGTGGTCAAACTGCTTAACAAAGCCATCAAAAAGCGAGGGGACTATGATGCCAACATCGTAGCTGTGGTGAATGACACAGTGGGCACCATGATGACCTGTGGCTATGACGACCAGCACTGTGAAGTCGGCCTGATCATCGGCACTGGCACCAATGCTTGCTACATGGAGGAACTGAGGCACATTGATCTGGTGGAAGGAGACGAGGGGAGGATGTGTATCAATACAGAATGGGGAGCCTTTGGAGACGATGGATCATTAGAAGACATCCGGACAGAGTTTGACAGGGAGATAGACCGGGGATCCCTCAACCCTGGAAAACAGCTGTTTGAGAAGATGGTCAGTGGCATGTACTTGGGAGAGCTGGTTCGACTGATCCTAGTCAAGATGGCCAAGGAGGGCCTCTTATTTGAAGGGCGGATCACCCCGGAGCTGCTCACCCGAGGGAAGTTTAACACCAGTGATGTGTCAGCCATCGAAAAGAATAAGGAAGGCCTCCACAATGCCAAAGAAATCCTGACCCGCCTGGGAGTGGAGCCGTCCGATGATGACTGTGTCTCAGTCCAGCACGTTTGCACCATTGTCTCATTTCGCTCAGCCAACTTGGTGGCTGCCACACTGGGCGCCATCTTGAACCGCCTGCGTGATAACAAGGGCACACCCAGGCTGCGGACCACGGTTGGTGTCGACGGATCTCTTTACAAGACGCACCCACAGTATTCCCGGCGTTTCCACAAGACTCTAAGGCGCTTGGTGCCAGACTCCGATGTGCGCTTCCTCCTCTCGGAGAGTGGCAGCGGCAAGGGGGCTGCCATGGTGACGGCGGTGGCCTACCGCTTGGCCGAGCAGCACCGGCAGATAGAGGAGACCCTGGCTCATTTCCACCTCACCAAGGACATGCTGCTGGAGGTGAAGAAGAGGATGCGGGCCGAGATGGAGCTGGGGCTGAGGAAGCAGACGCACAACAATGCCGTGGTTAAGATGCTGCCCTCCTTCGTCCGGAGAACTCCCGACGGGACCGAGAATGGTGACTTCTTGGCCCTGGATCTTGGAGGAACCAATTTCCGTGTGCTGCTGGTGAAAATCCGTAGTGGGAAAAAGAGAACGGTGGAAATGCACAACAAGATCTACGCCATTCCTATTGAAATCATGCAGGGCACTGGGGAAGAGCTGTTTGATCACATTGTCTCCTGCATCTCTGACTTCTTGGACTACATGGGGATCAAAGGCCCCAGGATGCCTCTGGGCTTCACGTTCTCATTTCCCTGCCAGCAGACGAGTCTGGACGCGGGAATCTTGATCACGTGGACAAAGGGTTTTAAGGCAACAGACTGCGTGGGCCACGATGTAGTCACCTTACTAAGGGATGCGATAAAAAGGAGAGAGGAATTTGACCTGGACGTGGTGGCTGTGGTCAACGACACAGTGGGCACCATGATGACCTGTGCTTATGAGGAGCCCACCTGTGAGGTTGGACTCATTGTTGGGACCGGCAGCAATGCCTGCTACATGGAGGAGATGAAGAACGTGGAGATGGTGGAGGGGGACCAGGGGCAGATGTGCATCAACATGGAGTGGGGGGCCTTTGGGGACAACGGGTGTCTGGATGATATCAGGACACACTACGACAGACTGGTGGACGAATATTCCCTAAATGCTGGGAAACAAAGGTATGAGAAGATGATCAGTGGTATGTACCTGGGTGAAATCGTCCGCAACATCTTAATCGACTTCACCAAGAAGGGATTCCTCTTCCGAGGGCAGATCTCTGAGACGCTGAAGACCCGGGGCATCTTTGAGACCAAGTTTCTCTCTCAGATCGAGAGTGACCGATTAGCACTGCTCCAGGTCCGGGCTATCCTCCAGCAGCTAGGTCTGAATAGCACCTGCGATGACAGTATCCTCGTCAAGACAGTGTGCGGGGTGGTGTCCAGGAGGGCCGCACAGCTGTGTGGCGCAGGCATGGCTGCGGTTGTGGATAAGATCCGCGAGAACAGAGGACTGGACCGTCTGAATGTGACTGTGGGAGTGGACGGGACACTCTACAAGCTTCATCCACACTTCTCCAGAATCATGCACCAGACGGTGAAGGAACTGTCACCAAAATGTAACGTGTCCTTCCTCCTGTCTGAGGATGGCAGCGGCAAGGGGGCCGCCCTCATCACGGCCGTGGGCGTGCGGTTACGCACAGAGGCAAGCAGCTAAGAGTCCGGGATCCCCAGCCTACTGCCTCTCCAGCACTTCTCTCTTCAAGCGGCGACCCCCTACCCTCCCAGCGAGTTGCGCTGGGAGACGCTGGCGCCAGGGCCTGCCGGCGCGGGGAGGAAAGCAAAATCCAACTAATGGTATATATTGTAGGGTACAGAATAGAGCGTGTGCTGTTGATAATATCTCTCACCCGGATCCCTCCTCACTTGCCCTGCCACTTTGCATGGTTTGATTTTGACCTGGTCCCCCACGTGTGAAGTGTAGTGGCATCCATTTCTAATGTATGCATTCATCCAACAGAGTTATTTATTGGCTGGAGATGGAAAATCACACCACCTGACAGGCCTTCTGGGCCTCCAAAGCCCATCCTTGGGGTTCCCCCTCCCTGTGTGAAATGTATTATCACCAGCAGACACTGCCGGGCCTCCCTCCCGGGGGCACTGCCTGAAGGCGAGTGTGGGCATAGCATTAGCTGCTTCCTCCCCTCCTGGCACCCACTGTGGCCTGGCATCGCATCGTGGTGTGTCAATGCCACAAAATCGTGTGTCCGTGGAACCAGTCCTAGCCGCGTGTGACAGTCTTGCATTCTGTTTGTCTCGTGGGGGGAGGTGGACAGTCCTGCGGAAATGTGTCTTGTCTCCATTTGGATAAAAGGAACCAACCAACAAACAATGCCATCACTGGAATTTCCCACCGCTTTGTGAGCCGTGTCGTATGACCTAGTAAACTTTGTACCAATTCAAA\n",
            "Template strand: TAGATAGAACGACACAAACCTGTCCGGTCGGGGACTTTGTAGAACCCGTTACCTCCCAATTGAAGAGTTTCAAATTATCCGTTCTGGTCGTTGGTACGTTGTTCCTGAAGTTGATTGGTTGATTTCTTGACAAGGGGTCTCGTAACAAGGACTCTTCCTTTTCTCAGGTTTGTGGATGGGTGTGGACGAAACACGGTTCTTAGGTGTCAACCTAACGTTCCTGTCACGCAAGTTCTGGGTCGACAACTCTCATCTTTTCGTCTTCTTTCCTGGGCTCCAGTCGTTCACGGGAGGGGTGTTACCCCGTCTAGACGGTCGCTCTTAGCCGATGTCGTCGACTTTTTGGTTTTGAAGTAGATGAACGACTTTCACTCTAACTGTTCATAGAGATACGGTACGCCGAGAGGCTACTTTGAGAGTATCTATAGTACTGAGCGAAGTCCTTCCTCTACTTCTTACCGGAGAGGGCCCTAAAATTAGGTTGTCGGTGTCAGTTCTACAACGGTTGTAAGCATTCCAGGTAAGGACTACCGAGACTTTTCCCTCTAAAGTAACGGGACCTAGAACCACCCAGAAGGAAAGCTTAAGACGCCCACGTTCACTTAGTACTCTTTTTGGTCTTACAAGTGTACCTCAGGCTCCAAATACTGTGGGGTCTCTTGTAGCACGTGCCGTCACCTTCGGTCGAAAAACTAGTACAACGACTCACGGACCCTCTAAAGTACCTCTTTTCCTTCTAGTTCCTGTTCTTCAATGGACACCCTAAGTGCAAAAGAAAAGGAACGGTTGTTAGGTTTTATCTACTCCGGTAGGACTAGTGGACCTGTTTCGCTAAATTTCGCTCGCCTCACCTTCCTCGTCTACACCAGTTTGACGAATTGTTTCGGTAGTTTTTCGCTCCCCTGATACTACGGTTGTAGCATCGACACCACTTACTGTGTCACCCGTGGTACTACTGGACACCGATACTGCTGGTCGTGACACTTCAGCCGGACTAGTAGCCGTGACCGTGGTTACGAACGATGTACCTCCTTGACTCCGTGTAACTAGACCACCTTCCTCTGCTCCCCTCCTACACATAGTTATGTCTTACCCCTCGGAAACCTCTGCTACCTAGTAATCTTCTGTAGGCCTGTCTCAAACTGTCCCTCTATCTGGCCCCTAGGGAGTTGGGACCTTTTGTCGACAAACTCTTCTACCAGTCACCGTACATGAACCCTCTCGACCAAGCTGACTAGGATCAGTTCTACCGGTTCCTCCCGGAGAATAAACTTCCCGCCTAGTGGGGCCTCGACGAGTGGGCTCCCTTCAAATTGTGGTCACTACACAGTCGGTAGCTTTTCTTATTCCTTCCGGAGGTGTTACGGTTTCTTTAGGACTGGGCGGACCCTCACCTCGGCAGGCTACTACTGACACAGAGTCAGGTCGTGCAAACGTGGTAACAGAGTAAAGCGAGTCGGTTGAACCACCGACGGTGTGACCCGCGGTAGAACTTGGCGGACGCACTATTGTTCCCGTGTGGGTCCGACGCCTGGTGCCAACCACAGCTGCCTAGAGAAATGTTCTGCGTGGGTGTCATAAGGGCCGCAAAGGTGTTCTGAGATTCCGCGAACCACGGTCTGAGGCTACACGCGAAGGAGGAGAGCCTCTCACCGTCGCCGTTCCCCCGACGGTACCACTGCCGCCACCGGATGGCGAACCGGCTCGTCGTGGCCGTCTATCTCCTCTGGGACCGAGTAAAGGTGGAGTGGTTCCTGTACGACGACCTCCACTTCTTCTCCTACGCCCGGCTCTACCTCGACCCCGACTCCTTCGTCTGCGTGTTGTTACGGCACCAATTCTACGACGGGAGGAAGCAGGCCTCTTGAGGGCTGCCCTGGCTCTTACCACTGAAGAACCGGGACCTAGAACCTCCTTGGTTAAAGGCACACGACGACCACTTTTAGGCATCACCCTTTTTCTCTTGCCACCTTTACGTGTTGTTCTAGATGCGGTAAGGATAACTTTAGTACGTCCCGTGACCCCTTCTCGACAAACTAGTGTAACAGAGGACGTAGAGACTGAAGAACCTGATGTACCCCTAGTTTCCGGGGTCCTACGGAGACCCGAAGTGCAAGAGTAAAGGGACGGTCGTCTGCTCAGACCTGCGCCCTTAGAACTAGTGCACCTGTTTCCCAAAATTCCGTTGTCTGACGCACCCGGTGCTACATCAGTGGAATGATTCCCTACGCTATTTTTCCTCTCTCCTTAAACTGGACCTGCACCACCGACACCAGTTGCTGTGTCACCCGTGGTACTACTGGACACGAATACTCCTCGGGTGGACACTCCAACCTGAGTAACAACCCTGGCCGTCGTTACGGACGATGTACCTCCTCTACTTCTTGCACCTCTACCACCTCCCCCTGGTCCCCGTCTACACGTAGTTGTACCTCACCCCCCGGAAACCCCTGTTGCCCACAGACCTACTATAGTCCTGTGTGATGCTGTCTGACCACCTGCTTATAAGGGATTTACGACCCTTTGTTTCCATACTCTTCTACTAGTCACCATACATGGACCCACTTTAGCAGGCGTTGTAGAATTAGCTGAAGTGGTTCTTCCCTAAGGAGAAGGCTCCCGTCTAGAGACTCTGCGACTTCTGGGCCCCGTAGAAACTCTGGTTCAAAGAGAGAGTCTAGCTCTCACTGGCTAATCGTGACGAGGTCCAGGCCCGATAGGAGGTCGTCGATCCAGACTTATCGTGGACGCTACTGTCATAGGAGCAGTTCTGTCACACGCCCCACCACAGGTCCTCCCGGCGTGTCGACACACCGCGTCCGTACCGACGCCAACACCTATTCTAGGCGCTCTTGTCTCCTGACCTGGCAGACTTACACTGACACCCTCACCTGCCCTGTGAGATGTTCGAAGTAGGTGTGAAGAGGTCTTAGTACGTGGTCTGCCACTTCCTTGACAGTGGTTTTACATTGCACAGGAAGGAGGACAGACTCCTACCGTCGCCGTTCCCCCGGCGGGAGTAGTGCCGGCACCCGCACGCCAATGCGTGTCTCCGTTCGTCGATTCTCAGGCCCTAGGGGTCGGATGACGGAGAGGTCGTGAAGAGAGAAGTTCGCCGCTGGGGGATGGGAGGGTCGCTCAACGCGACCCTCTGCGACCGCGGTCCCGGACGGCCGCGCCCCTCCTTTCGTTTTAGGTTGATTACCATATATAACATCCCATGTCTTATCTCGCACACGACAACTATTATAGAGAGTGGGCCTAGGGAGGAGTGAACGGGACGGTGAAACGTACCAAACTAAAACTGGACCAGGGGGTGCACACTTCACATCACCGTAGGTAAAGATTACATACGTAAGTAGGTTGTCTCAATAAATAACCGACCTCTACCTTTTAGTGTGGTGGACTGTCCGGAAGACCCGGAGGTTTCGGGTAGGAACCCCAAGGGGGAGGGACACACTTTACATAATAGTGGTCGTCTGTGACGGCCCGGAGGGAGGGCCCCCGTGACGGACTTCCGCTCACACCCGTATCGTAATCGACGAAGGAGGGGAGGACCGTGGGTGACACCGGACCGTAGCGTAGCACCACACAGTTACGGTGTTTTAGCACACAGGCACCTTGGTCAGGATCGGCGCACACTGTCAGAACGTAAGACAAACAGAGCACCCCCCTCCACCTGTCAGGACGCCTTTACACAGAACAGAGGTAAACCTATTTTCCTTGGTTGGTTGTTTGTTACGGTAGTGACCTTAAAGGGTGGCGAAACACTCGGCACAGCATACTGGATCATTTGAAACATGGTTAAGTTT\n",
            "mRNA: AUCUAUCUUGCUGUGUUUGGACAGGCCAGCCCCUGAAACAUCUUGGGCAAUGGAGGGUUAACUUCUCAAAGUUUAAUAGGCAAGACCAGCAACCAUGCAACAAGGACUUCAACUAACCAACUAAAGAACUGUUCCCCAGAGCAUUGUUCCUGAGAAGGAAAAGAGUCCAAACACCUACCCACACCUGCUUUGUGCCAAGAAUCCACAGUUGGAUUGCAAGGACAGUGCGUUCAAGACCCAGCUGUUGAGAGUAGAAAAGCAGAAGAAAGGACCCGAGGUCAGCAAGUGCCCUCCCCACAAUGGGGCAGAUCUGCCAGCGAGAAUCGGCUACAGCAGCUGAAAAACCAAAACUUCAUCUACUUGCUGAAAGUGAGAUUGACAAGUAUCUCUAUGCCAUGCGGCUCUCCGAUGAAACUCUCAUAGAUAUCAUGACUCGCUUCAGGAAGGAGAUGAAGAAUGGCCUCUCCCGGGAUUUUAAUCCAACAGCCACAGUCAAGAUGUUGCCAACAUUCGUAAGGUCCAUUCCUGAUGGCUCUGAAAAGGGAGAUUUCAUUGCCCUGGAUCUUGGUGGGUCUUCCUUUCGAAUUCUGCGGGUGCAAGUGAAUCAUGAGAAAAACCAGAAUGUUCACAUGGAGUCCGAGGUUUAUGACACCCCAGAGAACAUCGUGCACGGCAGUGGAAGCCAGCUUUUUGAUCAUGUUGCUGAGUGCCUGGGAGAUUUCAUGGAGAAAAGGAAGAUCAAGGACAAGAAGUUACCUGUGGGAUUCACGUUUUCUUUUCCUUGCCAACAAUCCAAAAUAGAUGAGGCCAUCCUGAUCACCUGGACAAAGCGAUUUAAAGCGAGCGGAGUGGAAGGAGCAGAUGUGGUCAAACUGCUUAACAAAGCCAUCAAAAAGCGAGGGGACUAUGAUGCCAACAUCGUAGCUGUGGUGAAUGACACAGUGGGCACCAUGAUGACCUGUGGCUAUGACGACCAGCACUGUGAAGUCGGCCUGAUCAUCGGCACUGGCACCAAUGCUUGCUACAUGGAGGAACUGAGGCACAUUGAUCUGGUGGAAGGAGACGAGGGGAGGAUGUGUAUCAAUACAGAAUGGGGAGCCUUUGGAGACGAUGGAUCAUUAGAAGACAUCCGGACAGAGUUUGACAGGGAGAUAGACCGGGGAUCCCUCAACCCUGGAAAACAGCUGUUUGAGAAGAUGGUCAGUGGCAUGUACUUGGGAGAGCUGGUUCGACUGAUCCUAGUCAAGAUGGCCAAGGAGGGCCUCUUAUUUGAAGGGCGGAUCACCCCGGAGCUGCUCACCCGAGGGAAGUUUAACACCAGUGAUGUGUCAGCCAUCGAAAAGAAUAAGGAAGGCCUCCACAAUGCCAAAGAAAUCCUGACCCGCCUGGGAGUGGAGCCGUCCGAUGAUGACUGUGUCUCAGUCCAGCACGUUUGCACCAUUGUCUCAUUUCGCUCAGCCAACUUGGUGGCUGCCACACUGGGCGCCAUCUUGAACCGCCUGCGUGAUAACAAGGGCACACCCAGGCUGCGGACCACGGUUGGUGUCGACGGAUCUCUUUACAAGACGCACCCACAGUAUUCCCGGCGUUUCCACAAGACUCUAAGGCGCUUGGUGCCAGACUCCGAUGUGCGCUUCCUCCUCUCGGAGAGUGGCAGCGGCAAGGGGGCUGCCAUGGUGACGGCGGUGGCCUACCGCUUGGCCGAGCAGCACCGGCAGAUAGAGGAGACCCUGGCUCAUUUCCACCUCACCAAGGACAUGCUGCUGGAGGUGAAGAAGAGGAUGCGGGCCGAGAUGGAGCUGGGGCUGAGGAAGCAGACGCACAACAAUGCCGUGGUUAAGAUGCUGCCCUCCUUCGUCCGGAGAACUCCCGACGGGACCGAGAAUGGUGACUUCUUGGCCCUGGAUCUUGGAGGAACCAAUUUCCGUGUGCUGCUGGUGAAAAUCCGUAGUGGGAAAAAGAGAACGGUGGAAAUGCACAACAAGAUCUACGCCAUUCCUAUUGAAAUCAUGCAGGGCACUGGGGAAGAGCUGUUUGAUCACAUUGUCUCCUGCAUCUCUGACUUCUUGGACUACAUGGGGAUCAAAGGCCCCAGGAUGCCUCUGGGCUUCACGUUCUCAUUUCCCUGCCAGCAGACGAGUCUGGACGCGGGAAUCUUGAUCACGUGGACAAAGGGUUUUAAGGCAACAGACUGCGUGGGCCACGAUGUAGUCACCUUACUAAGGGAUGCGAUAAAAAGGAGAGAGGAAUUUGACCUGGACGUGGUGGCUGUGGUCAACGACACAGUGGGCACCAUGAUGACCUGUGCUUAUGAGGAGCCCACCUGUGAGGUUGGACUCAUUGUUGGGACCGGCAGCAAUGCCUGCUACAUGGAGGAGAUGAAGAACGUGGAGAUGGUGGAGGGGGACCAGGGGCAGAUGUGCAUCAACAUGGAGUGGGGGGCCUUUGGGGACAACGGGUGUCUGGAUGAUAUCAGGACACACUACGACAGACUGGUGGACGAAUAUUCCCUAAAUGCUGGGAAACAAAGGUAUGAGAAGAUGAUCAGUGGUAUGUACCUGGGUGAAAUCGUCCGCAACAUCUUAAUCGACUUCACCAAGAAGGGAUUCCUCUUCCGAGGGCAGAUCUCUGAGACGCUGAAGACCCGGGGCAUCUUUGAGACCAAGUUUCUCUCUCAGAUCGAGAGUGACCGAUUAGCACUGCUCCAGGUCCGGGCUAUCCUCCAGCAGCUAGGUCUGAAUAGCACCUGCGAUGACAGUAUCCUCGUCAAGACAGUGUGCGGGGUGGUGUCCAGGAGGGCCGCACAGCUGUGUGGCGCAGGCAUGGCUGCGGUUGUGGAUAAGAUCCGCGAGAACAGAGGACUGGACCGUCUGAAUGUGACUGUGGGAGUGGACGGGACACUCUACAAGCUUCAUCCACACUUCUCCAGAAUCAUGCACCAGACGGUGAAGGAACUGUCACCAAAAUGUAACGUGUCCUUCCUCCUGUCUGAGGAUGGCAGCGGCAAGGGGGCCGCCCUCAUCACGGCCGUGGGCGUGCGGUUACGCACAGAGGCAAGCAGCUAAGAGUCCGGGAUCCCCAGCCUACUGCCUCUCCAGCACUUCUCUCUUCAAGCGGCGACCCCCUACCCUCCCAGCGAGUUGCGCUGGGAGACGCUGGCGCCAGGGCCUGCCGGCGCGGGGAGGAAAGCAAAAUCCAACUAAUGGUAUAUAUUGUAGGGUACAGAAUAGAGCGUGUGCUGUUGAUAAUAUCUCUCACCCGGAUCCCUCCUCACUUGCCCUGCCACUUUGCAUGGUUUGAUUUUGACCUGGUCCCCCACGUGUGAAGUGUAGUGGCAUCCAUUUCUAAUGUAUGCAUUCAUCCAACAGAGUUAUUUAUUGGCUGGAGAUGGAAAAUCACACCACCUGACAGGCCUUCUGGGCCUCCAAAGCCCAUCCUUGGGGUUCCCCCUCCCUGUGUGAAAUGUAUUAUCACCAGCAGACACUGCCGGGCCUCCCUCCCGGGGGCACUGCCUGAAGGCGAGUGUGGGCAUAGCAUUAGCUGCUUCCUCCCCUCCUGGCACCCACUGUGGCCUGGCAUCGCAUCGUGGUGUGUCAAUGCCACAAAAUCGUGUGUCCGUGGAACCAGUCCUAGCCGCGUGUGACAGUCUUGCAUUCUGUUUGUCUCGUGGGGGGAGGUGGACAGUCCUGCGGAAAUGUGUCUUGUCUCCAUUUGGAUAAAAGGAACCAACCAACAAACAAUGCCAUCACUGGAAUUUCCCACCGCUUUGUGAGCCGUGUCGUAUGACCUAGUAAACUUUGUACCAAUUCAAA\n",
            "Reverse complement: TTTGAATTGGTACAAAGTTTACTAGGTCATACGACACGGCTCACAAAGCGGTGGGAAATTCCAGTGATGGCATTGTTTGTTGGTTGGTTCCTTTTATCCAAATGGAGACAAGACACATTTCCGCAGGACTGTCCACCTCCCCCCACGAGACAAACAGAATGCAAGACTGTCACACGCGGCTAGGACTGGTTCCACGGACACACGATTTTGTGGCATTGACACACCACGATGCGATGCCAGGCCACAGTGGGTGCCAGGAGGGGAGGAAGCAGCTAATGCTATGCCCACACTCGCCTTCAGGCAGTGCCCCCGGGAGGGAGGCCCGGCAGTGTCTGCTGGTGATAATACATTTCACACAGGGAGGGGGAACCCCAAGGATGGGCTTTGGAGGCCCAGAAGGCCTGTCAGGTGGTGTGATTTTCCATCTCCAGCCAATAAATAACTCTGTTGGATGAATGCATACATTAGAAATGGATGCCACTACACTTCACACGTGGGGGACCAGGTCAAAATCAAACCATGCAAAGTGGCAGGGCAAGTGAGGAGGGATCCGGGTGAGAGATATTATCAACAGCACACGCTCTATTCTGTACCCTACAATATATACCATTAGTTGGATTTTGCTTTCCTCCCCGCGCCGGCAGGCCCTGGCGCCAGCGTCTCCCAGCGCAACTCGCTGGGAGGGTAGGGGGTCGCCGCTTGAAGAGAGAAGTGCTGGAGAGGCAGTAGGCTGGGGATCCCGGACTCTTAGCTGCTTGCCTCTGTGCGTAACCGCACGCCCACGGCCGTGATGAGGGCGGCCCCCTTGCCGCTGCCATCCTCAGACAGGAGGAAGGACACGTTACATTTTGGTGACAGTTCCTTCACCGTCTGGTGCATGATTCTGGAGAAGTGTGGATGAAGCTTGTAGAGTGTCCCGTCCACTCCCACAGTCACATTCAGACGGTCCAGTCCTCTGTTCTCGCGGATCTTATCCACAACCGCAGCCATGCCTGCGCCACACAGCTGTGCGGCCCTCCTGGACACCACCCCGCACACTGTCTTGACGAGGATACTGTCATCGCAGGTGCTATTCAGACCTAGCTGCTGGAGGATAGCCCGGACCTGGAGCAGTGCTAATCGGTCACTCTCGATCTGAGAGAGAAACTTGGTCTCAAAGATGCCCCGGGTCTTCAGCGTCTCAGAGATCTGCCCTCGGAAGAGGAATCCCTTCTTGGTGAAGTCGATTAAGATGTTGCGGACGATTTCACCCAGGTACATACCACTGATCATCTTCTCATACCTTTGTTTCCCAGCATTTAGGGAATATTCGTCCACCAGTCTGTCGTAGTGTGTCCTGATATCATCCAGACACCCGTTGTCCCCAAAGGCCCCCCACTCCATGTTGATGCACATCTGCCCCTGGTCCCCCTCCACCATCTCCACGTTCTTCATCTCCTCCATGTAGCAGGCATTGCTGCCGGTCCCAACAATGAGTCCAACCTCACAGGTGGGCTCCTCATAAGCACAGGTCATCATGGTGCCCACTGTGTCGTTGACCACAGCCACCACGTCCAGGTCAAATTCCTCTCTCCTTTTTATCGCATCCCTTAGTAAGGTGACTACATCGTGGCCCACGCAGTCTGTTGCCTTAAAACCCTTTGTCCACGTGATCAAGATTCCCGCGTCCAGACTCGTCTGCTGGCAGGGAAATGAGAACGTGAAGCCCAGAGGCATCCTGGGGCCTTTGATCCCCATGTAGTCCAAGAAGTCAGAGATGCAGGAGACAATGTGATCAAACAGCTCTTCCCCAGTGCCCTGCATGATTTCAATAGGAATGGCGTAGATCTTGTTGTGCATTTCCACCGTTCTCTTTTTCCCACTACGGATTTTCACCAGCAGCACACGGAAATTGGTTCCTCCAAGATCCAGGGCCAAGAAGTCACCATTCTCGGTCCCGTCGGGAGTTCTCCGGACGAAGGAGGGCAGCATCTTAACCACGGCATTGTTGTGCGTCTGCTTCCTCAGCCCCAGCTCCATCTCGGCCCGCATCCTCTTCTTCACCTCCAGCAGCATGTCCTTGGTGAGGTGGAAATGAGCCAGGGTCTCCTCTATCTGCCGGTGCTGCTCGGCCAAGCGGTAGGCCACCGCCGTCACCATGGCAGCCCCCTTGCCGCTGCCACTCTCCGAGAGGAGGAAGCGCACATCGGAGTCTGGCACCAAGCGCCTTAGAGTCTTGTGGAAACGCCGGGAATACTGTGGGTGCGTCTTGTAAAGAGATCCGTCGACACCAACCGTGGTCCGCAGCCTGGGTGTGCCCTTGTTATCACGCAGGCGGTTCAAGATGGCGCCCAGTGTGGCAGCCACCAAGTTGGCTGAGCGAAATGAGACAATGGTGCAAACGTGCTGGACTGAGACACAGTCATCATCGGACGGCTCCACTCCCAGGCGGGTCAGGATTTCTTTGGCATTGTGGAGGCCTTCCTTATTCTTTTCGATGGCTGACACATCACTGGTGTTAAACTTCCCTCGGGTGAGCAGCTCCGGGGTGATCCGCCCTTCAAATAAGAGGCCCTCCTTGGCCATCTTGACTAGGATCAGTCGAACCAGCTCTCCCAAGTACATGCCACTGACCATCTTCTCAAACAGCTGTTTTCCAGGGTTGAGGGATCCCCGGTCTATCTCCCTGTCAAACTCTGTCCGGATGTCTTCTAATGATCCATCGTCTCCAAAGGCTCCCCATTCTGTATTGATACACATCCTCCCCTCGTCTCCTTCCACCAGATCAATGTGCCTCAGTTCCTCCATGTAGCAAGCATTGGTGCCAGTGCCGATGATCAGGCCGACTTCACAGTGCTGGTCGTCATAGCCACAGGTCATCATGGTGCCCACTGTGTCATTCACCACAGCTACGATGTTGGCATCATAGTCCCCTCGCTTTTTGATGGCTTTGTTAAGCAGTTTGACCACATCTGCTCCTTCCACTCCGCTCGCTTTAAATCGCTTTGTCCAGGTGATCAGGATGGCCTCATCTATTTTGGATTGTTGGCAAGGAAAAGAAAACGTGAATCCCACAGGTAACTTCTTGTCCTTGATCTTCCTTTTCTCCATGAAATCTCCCAGGCACTCAGCAACATGATCAAAAAGCTGGCTTCCACTGCCGTGCACGATGTTCTCTGGGGTGTCATAAACCTCGGACTCCATGTGAACATTCTGGTTTTTCTCATGATTCACTTGCACCCGCAGAATTCGAAAGGAAGACCCACCAAGATCCAGGGCAATGAAATCTCCCTTTTCAGAGCCATCAGGAATGGACCTTACGAATGTTGGCAACATCTTGACTGTGGCTGTTGGATTAAAATCCCGGGAGAGGCCATTCTTCATCTCCTTCCTGAAGCGAGTCATGATATCTATGAGAGTTTCATCGGAGAGCCGCATGGCATAGAGATACTTGTCAATCTCACTTTCAGCAAGTAGATGAAGTTTTGGTTTTTCAGCTGCTGTAGCCGATTCTCGCTGGCAGATCTGCCCCATTGTGGGGAGGGCACTTGCTGACCTCGGGTCCTTTCTTCTGCTTTTCTACTCTCAACAGCTGGGTCTTGAACGCACTGTCCTTGCAATCCAACTGTGGATTCTTGGCACAAAGCAGGTGTGGGTAGGTGTTTGGACTCTTTTCCTTCTCAGGAACAATGCTCTGGGGAACAGTTCTTTAGTTGGTTAGTTGAAGTCCTTGTTGCATGGTTGCTGGTCTTGCCTATTAAACTTTGAGAAGTTAACCCTCCATTGCCCAAGATGTTTCAGGGGCTGGCCTGTCCAAACACAGCAAGATAGAT\n",
            "Reverse complement RNA: UUUGAAUUGGUACAAAGUUUACUAGGUCAUACGACACGGCUCACAAAGCGGUGGGAAAUUCCAGUGAUGGCAUUGUUUGUUGGUUGGUUCCUUUUAUCCAAAUGGAGACAAGACACAUUUCCGCAGGACUGUCCACCUCCCCCCACGAGACAAACAGAAUGCAAGACUGUCACACGCGGCUAGGACUGGUUCCACGGACACACGAUUUUGUGGCAUUGACACACCACGAUGCGAUGCCAGGCCACAGUGGGUGCCAGGAGGGGAGGAAGCAGCUAAUGCUAUGCCCACACUCGCCUUCAGGCAGUGCCCCCGGGAGGGAGGCCCGGCAGUGUCUGCUGGUGAUAAUACAUUUCACACAGGGAGGGGGAACCCCAAGGAUGGGCUUUGGAGGCCCAGAAGGCCUGUCAGGUGGUGUGAUUUUCCAUCUCCAGCCAAUAAAUAACUCUGUUGGAUGAAUGCAUACAUUAGAAAUGGAUGCCACUACACUUCACACGUGGGGGACCAGGUCAAAAUCAAACCAUGCAAAGUGGCAGGGCAAGUGAGGAGGGAUCCGGGUGAGAGAUAUUAUCAACAGCACACGCUCUAUUCUGUACCCUACAAUAUAUACCAUUAGUUGGAUUUUGCUUUCCUCCCCGCGCCGGCAGGCCCUGGCGCCAGCGUCUCCCAGCGCAACUCGCUGGGAGGGUAGGGGGUCGCCGCUUGAAGAGAGAAGUGCUGGAGAGGCAGUAGGCUGGGGAUCCCGGACUCUUAGCUGCUUGCCUCUGUGCGUAACCGCACGCCCACGGCCGUGAUGAGGGCGGCCCCCUUGCCGCUGCCAUCCUCAGACAGGAGGAAGGACACGUUACAUUUUGGUGACAGUUCCUUCACCGUCUGGUGCAUGAUUCUGGAGAAGUGUGGAUGAAGCUUGUAGAGUGUCCCGUCCACUCCCACAGUCACAUUCAGACGGUCCAGUCCUCUGUUCUCGCGGAUCUUAUCCACAACCGCAGCCAUGCCUGCGCCACACAGCUGUGCGGCCCUCCUGGACACCACCCCGCACACUGUCUUGACGAGGAUACUGUCAUCGCAGGUGCUAUUCAGACCUAGCUGCUGGAGGAUAGCCCGGACCUGGAGCAGUGCUAAUCGGUCACUCUCGAUCUGAGAGAGAAACUUGGUCUCAAAGAUGCCCCGGGUCUUCAGCGUCUCAGAGAUCUGCCCUCGGAAGAGGAAUCCCUUCUUGGUGAAGUCGAUUAAGAUGUUGCGGACGAUUUCACCCAGGUACAUACCACUGAUCAUCUUCUCAUACCUUUGUUUCCCAGCAUUUAGGGAAUAUUCGUCCACCAGUCUGUCGUAGUGUGUCCUGAUAUCAUCCAGACACCCGUUGUCCCCAAAGGCCCCCCACUCCAUGUUGAUGCACAUCUGCCCCUGGUCCCCCUCCACCAUCUCCACGUUCUUCAUCUCCUCCAUGUAGCAGGCAUUGCUGCCGGUCCCAACAAUGAGUCCAACCUCACAGGUGGGCUCCUCAUAAGCACAGGUCAUCAUGGUGCCCACUGUGUCGUUGACCACAGCCACCACGUCCAGGUCAAAUUCCUCUCUCCUUUUUAUCGCAUCCCUUAGUAAGGUGACUACAUCGUGGCCCACGCAGUCUGUUGCCUUAAAACCCUUUGUCCACGUGAUCAAGAUUCCCGCGUCCAGACUCGUCUGCUGGCAGGGAAAUGAGAACGUGAAGCCCAGAGGCAUCCUGGGGCCUUUGAUCCCCAUGUAGUCCAAGAAGUCAGAGAUGCAGGAGACAAUGUGAUCAAACAGCUCUUCCCCAGUGCCCUGCAUGAUUUCAAUAGGAAUGGCGUAGAUCUUGUUGUGCAUUUCCACCGUUCUCUUUUUCCCACUACGGAUUUUCACCAGCAGCACACGGAAAUUGGUUCCUCCAAGAUCCAGGGCCAAGAAGUCACCAUUCUCGGUCCCGUCGGGAGUUCUCCGGACGAAGGAGGGCAGCAUCUUAACCACGGCAUUGUUGUGCGUCUGCUUCCUCAGCCCCAGCUCCAUCUCGGCCCGCAUCCUCUUCUUCACCUCCAGCAGCAUGUCCUUGGUGAGGUGGAAAUGAGCCAGGGUCUCCUCUAUCUGCCGGUGCUGCUCGGCCAAGCGGUAGGCCACCGCCGUCACCAUGGCAGCCCCCUUGCCGCUGCCACUCUCCGAGAGGAGGAAGCGCACAUCGGAGUCUGGCACCAAGCGCCUUAGAGUCUUGUGGAAACGCCGGGAAUACUGUGGGUGCGUCUUGUAAAGAGAUCCGUCGACACCAACCGUGGUCCGCAGCCUGGGUGUGCCCUUGUUAUCACGCAGGCGGUUCAAGAUGGCGCCCAGUGUGGCAGCCACCAAGUUGGCUGAGCGAAAUGAGACAAUGGUGCAAACGUGCUGGACUGAGACACAGUCAUCAUCGGACGGCUCCACUCCCAGGCGGGUCAGGAUUUCUUUGGCAUUGUGGAGGCCUUCCUUAUUCUUUUCGAUGGCUGACACAUCACUGGUGUUAAACUUCCCUCGGGUGAGCAGCUCCGGGGUGAUCCGCCCUUCAAAUAAGAGGCCCUCCUUGGCCAUCUUGACUAGGAUCAGUCGAACCAGCUCUCCCAAGUACAUGCCACUGACCAUCUUCUCAAACAGCUGUUUUCCAGGGUUGAGGGAUCCCCGGUCUAUCUCCCUGUCAAACUCUGUCCGGAUGUCUUCUAAUGAUCCAUCGUCUCCAAAGGCUCCCCAUUCUGUAUUGAUACACAUCCUCCCCUCGUCUCCUUCCACCAGAUCAAUGUGCCUCAGUUCCUCCAUGUAGCAAGCAUUGGUGCCAGUGCCGAUGAUCAGGCCGACUUCACAGUGCUGGUCGUCAUAGCCACAGGUCAUCAUGGUGCCCACUGUGUCAUUCACCACAGCUACGAUGUUGGCAUCAUAGUCCCCUCGCUUUUUGAUGGCUUUGUUAAGCAGUUUGACCACAUCUGCUCCUUCCACUCCGCUCGCUUUAAAUCGCUUUGUCCAGGUGAUCAGGAUGGCCUCAUCUAUUUUGGAUUGUUGGCAAGGAAAAGAAAACGUGAAUCCCACAGGUAACUUCUUGUCCUUGAUCUUCCUUUUCUCCAUGAAAUCUCCCAGGCACUCAGCAACAUGAUCAAAAAGCUGGCUUCCACUGCCGUGCACGAUGUUCUCUGGGGUGUCAUAAACCUCGGACUCCAUGUGAACAUUCUGGUUUUUCUCAUGAUUCACUUGCACCCGCAGAAUUCGAAAGGAAGACCCACCAAGAUCCAGGGCAAUGAAAUCUCCCUUUUCAGAGCCAUCAGGAAUGGACCUUACGAAUGUUGGCAACAUCUUGACUGUGGCUGUUGGAUUAAAAUCCCGGGAGAGGCCAUUCUUCAUCUCCUUCCUGAAGCGAGUCAUGAUAUCUAUGAGAGUUUCAUCGGAGAGCCGCAUGGCAUAGAGAUACUUGUCAAUCUCACUUUCAGCAAGUAGAUGAAGUUUUGGUUUUUCAGCUGCUGUAGCCGAUUCUCGCUGGCAGAUCUGCCCCAUUGUGGGGAGGGCACUUGCUGACCUCGGGUCCUUUCUUCUGCUUUUCUACUCUCAACAGCUGGGUCUUGAACGCACUGUCCUUGCAAUCCAACUGUGGAUUCUUGGCACAAAGCAGGUGUGGGUAGGUGUUUGGACUCUUUUCCUUCUCAGGAACAAUGCUCUGGGGAACAGUUCUUUAGUUGGUUAGUUGAAGUCCUUGUUGCAUGGUUGCUGGUCUUGCCUAUUAAACUUUGAGAAGUUAACCCUCCAUUGCCCAAGAUGUUUCAGGGGCUGGCCUGUCCAAACACAGCAAGAUAGAU\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Exercise1\n",
        "\n",
        "a. Design forward and reverse primers for the Hexokinase gene, each 20 bp in length.\n",
        "\n",
        "b. Download the protein sequence of Hexokinase and extract its C-terminal region.\n",
        "\n",
        "Assignment:\n",
        "\n",
        "Locate the Hexokinase gene on Human Chromosome 10.\n",
        "\n",
        "Extract the gene sequence and compare it with the downloaded FASTA file.."
      ],
      "metadata": {
        "id": "w2gBeIXmU-U7"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dummy=hxk.seq[:20] # Extract the first 20 nucleotides of the Hexokinase gene sequence\n",
        "reverse_primer= (\"Reverse complemnet[:20]\")  # Get the reverse complement of the extracted sequence\n"
      ],
      "metadata": {
        "id": "j6SpIa7sAorL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import Entrez\n",
        "from Bio import SeqIO\n",
        "Entrez.email= 'EXAMPLE@gmail.com' # Set your email for NCBI Entrez\n",
        "handle=Entrez.efetch(db='protein', id='NP_277032.1', rettype='fasta', retmode='text')# Fetch the protein sequence for Hexokinase from the NCBI database\n",
        "hxk_prot=SeqIO.read(handle,'fasta') # Read the fetched data into a SeqRecord object\n",
        "print(hxk_prot)# Print the full protein sequence record\n",
        "hxk_prot.seq[225:548] # Extract and view a specific region of the protein sequence (C-terminal region)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "llbDRsu4VxiQ",
        "outputId": "456ca2da-b180-4c49-e4ef-7dcacf723c8c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ID: NP_277032.1\n",
            "Name: NP_277032.1\n",
            "Description: NP_277032.1 hexokinase-1 isoform HKI-ta/tb [Homo sapiens]\n",
            "Number of features: 0\n",
            "Seq('MGQICQRESATAAEKPKLHLLAESEIDKYLYAMRLSDETLIDIMTRFRKEMKNG...ASS')\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Seq('QHCEVGLIIGTGTNACYMEELRHIDLVEGDEGRMCINTEWGAFGDDGSLEDIRT...LVK')"
            ]
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Let's align some sequences (LOCAL ALIGNMENT)\n",
        "\n",
        "Tasks for Sequence Alignment\n",
        "\n",
        "1)Do pairwise alignments:\n",
        "Compare two gene or protein sequences to find the best match, both for the whole sequence (global) and for parts of it (local).\n",
        "\n",
        "2)Do multiple sequence alignment:\n",
        "Compare three or more sequences at the same time to see how they are similar or different.\n"
      ],
      "metadata": {
        "id": "X3_WutsGkcOl"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "BOOPj-dOrdlQ"
      },
      "outputs": [],
      "source": [
        "from Bio.Seq import Seq\n",
        "\n",
        "# Define DNA sequences using Biopython's Seq class\n",
        "seq1 = Seq(\"TGTGACTA\") # First DNA sequence\n",
        "seq2 = Seq(\"TGGCA\")   # Second DNA sequence\n",
        "seq3 = Seq(\"CATGGTCA\") # Third DNA sequence"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Pairwise Alignment Example\n",
        "\n",
        "This code uses Biopython's `Align` module to perform pairwise alignment between two DNA sequences:\n",
        "\n",
        "- **Aligner Creation**: A `PairwiseAligner` object is instantiated to align sequences.\n",
        "- **Performing Alignment**: The aligner is used to align the previously defined DNA sequences, `seq1` (\"TGTGACTA\") and `seq2` (\"TGGCA\").\n",
        "- **Output of Alignments**:alignment result is printed, showing how the two sequences match.\n",
        "- **Alignment Score**: The score of the alignment is calculated and printed, indicating the quality of the alignment based on the default scoring scheme.\n",
        "\n"
      ],
      "metadata": {
        "id": "bWVSem5INkgl"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import Align\n",
        "aligner = Align.PairwiseAligner()\n",
        "print(aligner)\n",
        "alignments = aligner.align(seq1, seq2)\n",
        "for alignment in alignments:\n",
        "   print(alignment)  #output of the alignement\n",
        "print('Score:', aligner.score(seq1,seq2)) #alignment score"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "c_fkU2Evru5T",
        "outputId": "9f652ced-0d24-453b-a19c-63f55334e85b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Pairwise sequence aligner with parameters\n",
            "  wildcard: None\n",
            "  match_score: 1.000000\n",
            "  mismatch_score: 0.000000\n",
            "  target_internal_open_gap_score: 0.000000\n",
            "  target_internal_extend_gap_score: 0.000000\n",
            "  target_left_open_gap_score: 0.000000\n",
            "  target_left_extend_gap_score: 0.000000\n",
            "  target_right_open_gap_score: 0.000000\n",
            "  target_right_extend_gap_score: 0.000000\n",
            "  query_internal_open_gap_score: 0.000000\n",
            "  query_internal_extend_gap_score: 0.000000\n",
            "  query_left_open_gap_score: 0.000000\n",
            "  query_left_extend_gap_score: 0.000000\n",
            "  query_right_open_gap_score: 0.000000\n",
            "  query_right_extend_gap_score: 0.000000\n",
            "  mode: global\n",
            "\n",
            "target            0 TGTGACTA 8\n",
            "                  0 ||-|-|-| 8\n",
            "query             0 TG-G-C-A 5\n",
            "\n",
            "Score: 5.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Percentage Identity Calculation\n",
        "\n",
        "below code defines a function to calculate the percentage identity between two aligned sequences i.e quantitative measure of how similar two sequences are after alignment.\n",
        "\n",
        "- **Function Definition**: The `percent_id` function takes three parameters:\n",
        "  - `score`: The alignment score obtained from the pairwise alignment.\n",
        "  - `seq1`: The first DNA sequence.\n",
        "  - `seq2`: The second DNA sequence.\n",
        "  \n",
        "- **Calculation**:\n",
        "  - The mean length of the two sequences is calculated.\n",
        "  - The percentage identity is computed using the formula given.\n",
        "\n",
        "\n",
        "  - The alignment score is obtained using the aligner.\n",
        "  - The percentage identity is calculated by calling the `percent_id` function.\n",
        "  - Finally, the percentage identity is printed.\n"
      ],
      "metadata": {
        "id": "5PA73Is-Oct2"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Let's write a function to calculate percentage identity the formula used\n",
        "def percent_id(score, seq1, seq2):\n",
        "  mean_length = (len(seq1) + len(seq2)) / 2\n",
        "  percentage_identity = (score / mean_length) * 100\n",
        "  return percentage_identity\n",
        "\n",
        "score= aligner.score(seq1,seq2) #alignment score\n",
        "pid= percent_id(score, seq1,seq2) #precentage identity\n",
        "print(\"Percentage identity:\", pid) #printed percentage"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TtvSdnPUsaRx",
        "outputId": "c8c08fb1-4930-48e8-ee87-d724989846a0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Percentage identity: 76.92307692307693\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Local Alignment with Custom Gap Scoring\n",
        "\n",
        "below code snippet demonstrates how to perform local alignment between two DNA sequences:\n",
        "\n",
        "- **Setting Alignment Mode**: The aligner mode is set to **local**, allowing the identification of the best matching region between the two sequences.\n",
        "  \n",
        "- **Gap Scoring**:\n",
        "  - **Open Gap Score**: Set to -1, indicating the penalty for opening a gap in the alignment.\n",
        "  - **Extend Gap Score**: Set to -0.5, representing the penalty for extending an existing gap.\n",
        "\n",
        "- **Aligner Information**: The aligner settings are printed to provide details about the current configuration.\n",
        "\n",
        "- **Performing Alignment**: The aligner is used to align the sequences `seq1` (\"TGTGACTA\") and `seq2` (\"TGGCA\").\n",
        "\n",
        "- **Output of Alignments**: Each alignment result is printed, showing how the two sequences match locally.\n",
        "\n",
        "- **Alignment Score**: The score of the local alignment is printed, providing insight into the quality of the alignment based on the specified scoring system.\n",
        "\n",
        "This example illustrates how to adjust gap penalties in local alignments to better fit specific sequence characteristics.\n"
      ],
      "metadata": {
        "id": "yqMVR8IPPQxn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#local alignmemt\n",
        "aligner.mode = \"local\"  #setting alignment mode\n",
        "aligner.open_gap_score = -1  #open gap score\n",
        "aligner.extend_gap_score = -0.5  #extend gap score\n",
        "print(aligner)                    # to get aligner information\n",
        "alignments = aligner.align(seq1, seq2) #aligner is used to align the two seq\n",
        "for alignment in alignments:\n",
        "    print(alignment)             #output of alignment is printed\n",
        "print(aligner.score(seq1,seq2)) #alignment score"
      ],
      "metadata": {
        "id": "ygwKwWsNyGJe",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e915508a-55b3-4c12-97b9-d6380e4862f1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Pairwise sequence aligner with parameters\n",
            "  wildcard: None\n",
            "  match_score: 1.000000\n",
            "  mismatch_score: 0.000000\n",
            "  target_internal_open_gap_score: -1.000000\n",
            "  target_internal_extend_gap_score: -0.500000\n",
            "  target_left_open_gap_score: -1.000000\n",
            "  target_left_extend_gap_score: -0.500000\n",
            "  target_right_open_gap_score: -1.000000\n",
            "  target_right_extend_gap_score: -0.500000\n",
            "  query_internal_open_gap_score: -1.000000\n",
            "  query_internal_extend_gap_score: -0.500000\n",
            "  query_left_open_gap_score: -1.000000\n",
            "  query_left_extend_gap_score: -0.500000\n",
            "  query_right_open_gap_score: -1.000000\n",
            "  query_right_extend_gap_score: -0.500000\n",
            "  mode: local\n",
            "\n",
            "target            0 TGTGA 5\n",
            "                  0 ||..| 5\n",
            "query             0 TGGCA 5\n",
            "\n",
            "target            2 TGAC 6\n",
            "                  0 ||.| 4\n",
            "query             0 TGGC 4\n",
            "\n",
            "3.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Loading the BLOSUM62 Substitution Matrix\n",
        "\n",
        "The below code snippet demonstrates how to load and display the BLOSUM62 substitution matrix: highlighting how to work with substitution matrice.\n",
        "\n",
        "- **Importing Substitution Matrices**: The `substitution_matrices` module from Biopython is imported to access various substitution matrices.\n",
        "\n",
        "- **Loading Substitution Matrices**: The `load()` function is called to load all available substitution matrices. This is useful for checking the available options.\n",
        "\n",
        "- **Loading BLOSUM62 Matrix**: Specifically, the BLOSUM62 substitution matrix is loaded using the `load(\"BLOSUM62\")` function. This matrix is commonly used for protein sequence alignment and provides scores for substituting one amino acid for another.\n",
        "\n",
        "- **Displaying the Matrix**: The BLOSUM62 matrix is printed, showing the substitution scores between amino acids. Positive values indicate favorable substitutions, while negative values indicate unfavorable ones.\n"
      ],
      "metadata": {
        "id": "k3ff1kJaQQMQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.Align import substitution_matrices #importing substitution matrix\n",
        "substitution_matrices.load()                #loading substitution matrix\n",
        "matrix = substitution_matrices.load(\"BLOSUM62\") #loading BLOSUM62 matrix\n",
        "print(matrix)  #displaying the matrix"
      ],
      "metadata": {
        "id": "7J_pRz3NgRop",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bdae46c4-4ab9-4e1b-a345-c5deb6e33a04"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "#  Matrix made by matblas from blosum62.iij\n",
            "#  * column uses minimum score\n",
            "#  BLOSUM Clustered Scoring Matrix in 1/2 Bit Units\n",
            "#  Blocks Database = /data/blocks_5.0/blocks.dat\n",
            "#  Cluster Percentage: >= 62\n",
            "#  Entropy =   0.6979, Expected =  -0.5209\n",
            "     A    R    N    D    C    Q    E    G    H    I    L    K    M    F    P    S    T    W    Y    V    B    Z    X    *\n",
            "A  4.0 -1.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  0.0 -3.0 -2.0  0.0 -2.0 -1.0  0.0 -4.0\n",
            "R -1.0  5.0  0.0 -2.0 -3.0  1.0  0.0 -2.0  0.0 -3.0 -2.0  2.0 -1.0 -3.0 -2.0 -1.0 -1.0 -3.0 -2.0 -3.0 -1.0  0.0 -1.0 -4.0\n",
            "N -2.0  0.0  6.0  1.0 -3.0  0.0  0.0  0.0  1.0 -3.0 -3.0  0.0 -2.0 -3.0 -2.0  1.0  0.0 -4.0 -2.0 -3.0  3.0  0.0 -1.0 -4.0\n",
            "D -2.0 -2.0  1.0  6.0 -3.0  0.0  2.0 -1.0 -1.0 -3.0 -4.0 -1.0 -3.0 -3.0 -1.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
            "C  0.0 -3.0 -3.0 -3.0  9.0 -3.0 -4.0 -3.0 -3.0 -1.0 -1.0 -3.0 -1.0 -2.0 -3.0 -1.0 -1.0 -2.0 -2.0 -1.0 -3.0 -3.0 -2.0 -4.0\n",
            "Q -1.0  1.0  0.0  0.0 -3.0  5.0  2.0 -2.0  0.0 -3.0 -2.0  1.0  0.0 -3.0 -1.0  0.0 -1.0 -2.0 -1.0 -2.0  0.0  3.0 -1.0 -4.0\n",
            "E -1.0  0.0  0.0  2.0 -4.0  2.0  5.0 -2.0  0.0 -3.0 -3.0  1.0 -2.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
            "G  0.0 -2.0  0.0 -1.0 -3.0 -2.0 -2.0  6.0 -2.0 -4.0 -4.0 -2.0 -3.0 -3.0 -2.0  0.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -1.0 -4.0\n",
            "H -2.0  0.0  1.0 -1.0 -3.0  0.0  0.0 -2.0  8.0 -3.0 -3.0 -1.0 -2.0 -1.0 -2.0 -1.0 -2.0 -2.0  2.0 -3.0  0.0  0.0 -1.0 -4.0\n",
            "I -1.0 -3.0 -3.0 -3.0 -1.0 -3.0 -3.0 -4.0 -3.0  4.0  2.0 -3.0  1.0  0.0 -3.0 -2.0 -1.0 -3.0 -1.0  3.0 -3.0 -3.0 -1.0 -4.0\n",
            "L -1.0 -2.0 -3.0 -4.0 -1.0 -2.0 -3.0 -4.0 -3.0  2.0  4.0 -2.0  2.0  0.0 -3.0 -2.0 -1.0 -2.0 -1.0  1.0 -4.0 -3.0 -1.0 -4.0\n",
            "K -1.0  2.0  0.0 -1.0 -3.0  1.0  1.0 -2.0 -1.0 -3.0 -2.0  5.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  0.0  1.0 -1.0 -4.0\n",
            "M -1.0 -1.0 -2.0 -3.0 -1.0  0.0 -2.0 -3.0 -2.0  1.0  2.0 -1.0  5.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0  1.0 -3.0 -1.0 -1.0 -4.0\n",
            "F -2.0 -3.0 -3.0 -3.0 -2.0 -3.0 -3.0 -3.0 -1.0  0.0  0.0 -3.0  0.0  6.0 -4.0 -2.0 -2.0  1.0  3.0 -1.0 -3.0 -3.0 -1.0 -4.0\n",
            "P -1.0 -2.0 -2.0 -1.0 -3.0 -1.0 -1.0 -2.0 -2.0 -3.0 -3.0 -1.0 -2.0 -4.0  7.0 -1.0 -1.0 -4.0 -3.0 -2.0 -2.0 -1.0 -2.0 -4.0\n",
            "S  1.0 -1.0  1.0  0.0 -1.0  0.0  0.0  0.0 -1.0 -2.0 -2.0  0.0 -1.0 -2.0 -1.0  4.0  1.0 -3.0 -2.0 -2.0  0.0  0.0  0.0 -4.0\n",
            "T  0.0 -1.0  0.0 -1.0 -1.0 -1.0 -1.0 -2.0 -2.0 -1.0 -1.0 -1.0 -1.0 -2.0 -1.0  1.0  5.0 -2.0 -2.0  0.0 -1.0 -1.0  0.0 -4.0\n",
            "W -3.0 -3.0 -4.0 -4.0 -2.0 -2.0 -3.0 -2.0 -2.0 -3.0 -2.0 -3.0 -1.0  1.0 -4.0 -3.0 -2.0 11.0  2.0 -3.0 -4.0 -3.0 -2.0 -4.0\n",
            "Y -2.0 -2.0 -2.0 -3.0 -2.0 -1.0 -2.0 -3.0  2.0 -1.0 -1.0 -2.0 -1.0  3.0 -3.0 -2.0 -2.0  2.0  7.0 -1.0 -3.0 -2.0 -1.0 -4.0\n",
            "V  0.0 -3.0 -3.0 -3.0 -1.0 -2.0 -2.0 -3.0 -3.0  3.0  1.0 -2.0  1.0 -1.0 -2.0 -2.0  0.0 -3.0 -1.0  4.0 -3.0 -2.0 -1.0 -4.0\n",
            "B -2.0 -1.0  3.0  4.0 -3.0  0.0  1.0 -1.0  0.0 -3.0 -4.0  0.0 -3.0 -3.0 -2.0  0.0 -1.0 -4.0 -3.0 -3.0  4.0  1.0 -1.0 -4.0\n",
            "Z -1.0  0.0  0.0  1.0 -3.0  3.0  4.0 -2.0  0.0 -3.0 -3.0  1.0 -1.0 -3.0 -1.0  0.0 -1.0 -3.0 -2.0 -2.0  1.0  4.0 -1.0 -4.0\n",
            "X  0.0 -1.0 -1.0 -1.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -1.0 -2.0  0.0  0.0 -2.0 -1.0 -1.0 -1.0 -1.0 -1.0 -4.0\n",
            "* -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0 -4.0  1.0\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Setting the Substitution Matrix for Alignment\n",
        "\n",
        "The below code snippet demonstrates how to set a custom substitution matrix for the aligner and calculate the alignment score using the BLOSUM62 matrix:\n",
        "\n",
        "- **Assigning the Substitution Matrix**: The previously loaded BLOSUM62 substitution matrix is assigned to the aligner with `aligner.substitution_matrix = matrix`. This allows the aligner to use the specific scoring scheme defined by the BLOSUM62 matrix for amino acid substitutions during the alignment process.\n",
        "\n",
        "- **Calculating the Alignment Score**: The score of the alignment between the sequences `seq1` (\"TGTGACTA\") and `seq2` (\"TGGCA\") is calculated using the aligner's scoring method. The resulting score reflects the quality of the alignment based on the provided substitution matrix."
      ],
      "metadata": {
        "id": "Ss3F4nhSRICM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "aligner.substitution_matrix = matrix\n",
        "print(aligner.score(seq1,seq2))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aIe65FuOi_ri",
        "outputId": "d4bdff12-17f7-45bd-ee44-e7f864e9dba6"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "27.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Exercise 2:\n",
        "Aligning Protein Sequences\n",
        "\n",
        "Retrieve the Protein Sequence: Download the protein sequence with the accession number NP_277035.2 from the NCBI database.\n",
        "\n",
        "Align with Hexokinase Protein: Use the previously retrieved Hexokinase protein sequence to perform an alignment with the newly downloaded protein sequence.\n",
        "\n",
        "Use Substitution Matrices: Apply an appropriate substitution matrix, such as BLOSUM62, during the alignment process."
      ],
      "metadata": {
        "id": "mYIVdXQPhyo-"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO\n",
        "for sequence in SeqIO.parse(\"/content/drive/MyDrive/EXAMPLE/all_hxks.fasta\",'fasta'):\n",
        "  hxk=sequence\n"
      ],
      "metadata": {
        "id": "zzqFCm_bJaB3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Slicing Sequences from Hexokinase Protein\n",
        "\n",
        "- **Sequence Definitions**:\n",
        "  - **`seq1`**: The first sequence is created by slicing the Hexokinase protein sequence (`hxk`) from the beginning to the 38th amino acid. This represents the first part of the protein.\n",
        "  - **`seq2`**: The second sequence is created by slicing the Hexokinase protein sequence from the 16th to the 29th amino acid, extracting a smaller region within the protein.\n",
        "\n",
        "- **Output**: Both sliced sequences are printed to the console, allowing for further analysis or comparison.\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "vSTVTF_5SRJ0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.Seq import Seq\n",
        "seq1 = hxk[:38] #slicing\n",
        "seq2 = hxk[16:29] #slicing\n",
        "print(seq1) #OUTPUT\n",
        "print(seq2) #OUTPUT\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fQtehbOkK_mK",
        "outputId": "2df7316b-7306-4aab-9d0d-7dcbff5becee"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('TGTGTTTGGACAG')\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "The below code snippet demonstrates how to retrieve a protein sequence from the NCBI database and compare its length with another protein sequence (Hexokinase):\n",
        "\n",
        "- **Assign Hexokinase Sequence**: The Hexokinase protein sequence (`hxk_prot`) is stored in the variable `hxk_1`.\n",
        "\n",
        "- **Fetch New Protein Sequence**:\n",
        "  - The protein sequence with the accession number **NP_277035.2** is retrieved from the NCBI database using the `Entrez.efetch` function.\n",
        "  - The retrieved data is read into a `SeqRecord` object using `SeqIO.read()`, and the sequence is stored in the variable `hxk_2`.\n",
        "\n",
        "- **Output Sequence Lengths**:\n",
        "  - The lengths of both protein sequences are printed using `print(len(hxk_1))` and `print(len(hxk_2))`, allowing for a quick comparison of their sizes.\n",
        "  \n",
        "- **Output Sequences**: The actual sequences are printed to the console, displaying the amino acid compositions of both proteins."
      ],
      "metadata": {
        "id": "kb8LVE9pSzuM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "hxk_1 = hxk_prot.seq #asigning hexokinase sequence in varible\n",
        "handle=Entrez.efetch(db='protein', id='NP_277035.2', rettype='fasta', retmode='text') #fetching new protien sequence into the variable handle\n",
        "hxk2_prot=SeqIO.read(handle,'fasta') #This line reads the data fetched from NCBI using SeqIO.read().\n",
        "                                      #The FASTA format is specified, and the resulting sequence data is stored in a SeqRecord object named hxk2_prot.\n",
        "hxk_2 = hxk2_prot.seq\n",
        "print(len(hxk_1))\n",
        "print(len(hxk_2))\n",
        "print(hxk_1)\n",
        "print(hxk_2)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aGd7rI2fhxPH",
        "outputId": "9a6d4e15-5c27-477a-966c-77e8eb94e010"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "921\n",
            "905\n",
            "MGQICQRESATAAEKPKLHLLAESEIDKYLYAMRLSDETLIDIMTRFRKEMKNGLSRDFNPTATVKMLPTFVRSIPDGSEKGDFIALDLGGSSFRILRVQVNHEKNQNVHMESEVYDTPENIVHGSGSQLFDHVAECLGDFMEKRKIKDKKLPVGFTFSFPCQQSKIDEAILITWTKRFKASGVEGADVVKLLNKAIKKRGDYDANIVAVVNDTVGTMMTCGYDDQHCEVGLIIGTGTNACYMEELRHIDLVEGDEGRMCINTEWGAFGDDGSLEDIRTEFDREIDRGSLNPGKQLFEKMVSGMYLGELVRLILVKMAKEGLLFEGRITPELLTRGKFNTSDVSAIEKNKEGLHNAKEILTRLGVEPSDDDCVSVQHVCTIVSFRSANLVAATLGAILNRLRDNKGTPRLRTTVGVDGSLYKTHPQYSRRFHKTLRRLVPDSDVRFLLSESGSGKGAAMVTAVAYRLAEQHRQIEETLAHFHLTKDMLLEVKKRMRAEMELGLRKQTHNNAVVKMLPSFVRRTPDGTENGDFLALDLGGTNFRVLLVKIRSGKKRTVEMHNKIYAIPIEIMQGTGEELFDHIVSCISDFLDYMGIKGPRMPLGFTFSFPCQQTSLDAGILITWTKGFKATDCVGHDVVTLLRDAIKRREEFDLDVVAVVNDTVGTMMTCAYEEPTCEVGLIVGTGSNACYMEEMKNVEMVEGDQGQMCINMEWGAFGDNGCLDDIRTHYDRLVDEYSLNAGKQRYEKMISGMYLGEIVRNILIDFTKKGFLFRGQISETLKTRGIFETKFLSQIESDRLALLQVRAILQQLGLNSTCDDSILVKTVCGVVSRRAAQLCGAGMAAVVDKIRENRGLDRLNVTVGVDGTLYKLHPHFSRIMHQTVKELSPKCNVSFLLSEDGSGKGAALITAVGVRLRTEASS\n",
            "MAKRALHDFIDKYLYAMRLSDETLIDIMTRFRKEMKNGLSRDFNPTATVKMLPTFVRSIPDGSEKGDFIALDLGGSSFRILRVQVNHEKNQNVHMESEVYDTPENIVHGSGSQLFDHVAECLGDFMEKRKIKDKKLPVGFTFSFPCQQSKIDEAILITWTKRFKASGVEGADVVKLLNKAIKKRGDYDANIVAVVNDTVGTMMTCGYDDQHCEVGLIIGTGTNACYMEELRHIDLVEGDEGRMCINTEWGAFGDDGSLEDIRTEFDREIDRGSLNPGKQLFEKMVSGMYLGELVRLILVKMAKEGLLFEGRITPELLTRGKFNTSDVSAIEKNKEGLHNAKEILTRLGVEPSDDDCVSVQHVCTIVSFRSANLVAATLGAILNRLRDNKGTPRLRTTVGVDGSLYKTHPQYSRRFHKTLRRLVPDSDVRFLLSESGSGKGAAMVTAVAYRLAEQHRQIEETLAHFHLTKDMLLEVKKRMRAEMELGLRKQTHNNAVVKMLPSFVRRTPDGTENGDFLALDLGGTNFRVLLVKIRSGKKRTVEMHNKIYAIPIEIMQGTGEELFDHIVSCISDFLDYMGIKGPRMPLGFTFSFPCQQTSLDAGILITWTKGFKATDCVGHDVVTLLRDAIKRREEFDLDVVAVVNDTVGTMMTCAYEEPTCEVGLIVGTGSNACYMEEMKNVEMVEGDQGQMCINMEWGAFGDNGCLDDIRTHYDRLVDEYSLNAGKQRYEKMISGMYLGEIVRNILIDFTKKGFLFRGQISETLKTRGIFETKFLSQIESDRLALLQVRAILQQLGLNSTCDDSILVKTVCGVVSRRAAQLCGAGMAAVVDKIRENRGLDRLNVTVGVDGTLYKLHPHFSRIMHQTVKELSPKCNVSFLLSEDGSGKGAALITAVGVRLRTEASS\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Protein Sequence Alignment and Analysi\n",
        "\n",
        "aligning the two protein sequences and calculating their alignment score and percentage identity:\n",
        "\n",
        "1. **Performing Protein Alignment**:\n",
        "   ```python\n",
        "   prot_align = aligner.align(hxk_1, hxk_2)\n"
      ],
      "metadata": {
        "id": "wVljI0e0UXIW"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "prot_align = aligner.align(hxk_1,hxk_2)\n",
        "#for alignment in prot_align:\n",
        "#    print(alignment)\n",
        "\n",
        "score= aligner.score(hxk_1,hxk_2)\n",
        "print(\"Score:\", score) #alignment score\n",
        "pid= percent_id(score, hxk_1,hxk_2)\n",
        "print(\"Percentage identity:\", pid) #percentage identity"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8nUUiXtdnLjS",
        "outputId": "a0772ee7-ddaa-43af-f33d-df93e68b0886"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Score: 4649.0\n",
            "Percentage identity: 509.20043811610077\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Another function for pairwise alignment\n",
        "\n",
        "1. **Importing the `pairwise2` Module**:\n",
        "\n",
        "   from Bio import pairwise2\n",
        "\n",
        "2. The pairwise2.align.globalxx() function is called to perform a global alignment of seq1 and seq3.\n",
        "3.The pairwise2.align.localxx() function performs a local alignment, identifying the most similar regions between seq1 and seq3 without aligning the entire sequences.\n",
        "4.The format_alignment() function is imported to format the alignment results for better readability.\n",
        "A loop iterates through each global alignment result, and the formatted alignments are printed to the console."
      ],
      "metadata": {
        "id": "2TakmCLBmym3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import pairwise2\n",
        "global_alignments = pairwise2.align.globalxx(seq1, seq3)\n",
        "print(global_alignments)\n",
        "local_alignments = pairwise2.align.localxx(seq1,seq3)\n",
        "\n",
        "from Bio.pairwise2 import format_alignment\n",
        "for alignment in global_alignments:\n",
        "    print(format_alignment(*alignment))\n",
        "\n",
        "print(\"\\n***************\\n\")  #A separator line is printed for clarity between global and local alignment results.\n",
        "\n",
        "for alignment in local_alignments:\n",
        "    print(format_alignment(*alignment))"
      ],
      "metadata": {
        "id": "93o3iACOiD1J",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "53ba4a5e-af11-418a-fbb4-fe6686ac3c8d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
            "  warnings.warn(\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "<Bio.Align.PairwiseAlignments object at 0x7a8455733d90>\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                         | |      || |  |        ||       \n",
            "---------------------------------------------------------------------------------------------------------------------------------------------------------C-A------TG-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      |    |      || |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------C----A------TG-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      ||          || |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------CA----------TG-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                         | |   |   | |  |        ||       \n",
            "---------------------------------------------------------------------------------------------------------------------------------------------------------C-A---T---G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      |    |   |   | |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------C----A---T---G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      ||       |   | |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------CA-------T---G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                         | |  |    | |  |        ||       \n",
            "---------------------------------------------------------------------------------------------------------------------------------------------------------C-A--T----G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      |    |  |    | |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------C----A--T----G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      ||      |    | |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------CA------T----G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                         | ||      | |  |        ||       \n",
            "---------------------------------------------------------------------------------------------------------------------------------------------------------C-AT------G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      |    ||      | |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------C----AT------G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      ||    |      | |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------CA----T------G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      ||  |        | |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------CA--T--------G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "ID: NM_033500.2\n",
            "Name: NM_033500.2\n",
            "Description: NM_033500.2 Homo sapiens hexokinase 1 (HK1), transcript variant 5, mRNA\n",
            "Number of features: 0\n",
            "Seq('AAAACATCTATCTTGCTGTGTTTGGACAGGCCAGCCCC')\n",
            "                                                                                                                                                      |||          | |  |        ||       \n",
            "------------------------------------------------------------------------------------------------------------------------------------------------------CAT----------G-G--T--------CA-------\n",
            "  Score=8\n",
            "\n",
            "\n",
            "***************\n",
            "\n",
            "154 CTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    | |      || |  |        ||\n",
            "  1 C-A------TG-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    |    |      || |  |        ||\n",
            "  1 C----A------TG-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    ||          || |  |        ||\n",
            "  1 CA----------TG-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "154 CTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    | |   |   | |  |        ||\n",
            "  1 C-A---T---G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    |    |   |   | |  |        ||\n",
            "  1 C----A---T---G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    ||       |   | |  |        ||\n",
            "  1 CA-------T---G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "154 CTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    | |  |    | |  |        ||\n",
            "  1 C-A--T----G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    |    |  |    | |  |        ||\n",
            "  1 C----A--T----G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    ||      |    | |  |        ||\n",
            "  1 CA------T----G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "154 CTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    | ||      | |  |        ||\n",
            "  1 C-AT------G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    |    ||      | |  |        ||\n",
            "  1 C----AT------G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    ||    |      | |  |        ||\n",
            "  1 CA----T------G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    ||  |        | |  |        ||\n",
            "  1 CA--T--------G-G--T--------CA\n",
            "  Score=8\n",
            "\n",
            "151 CATCTATCTTGCTGTGTTTGGACAGGCCA\n",
            "    |||          | |  |        ||\n",
            "  1 CAT----------G-G--T--------CA\n",
            "  Score=8\n",
            "\n"
          ]
        }
      ]
    }
  ]
}